{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Generating training instances from genomic dna, intron and mRNA positions.\n",
    "\n",
    "> Preparing training instances."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Goals\n",
    "1. Construct the mRNA sequence\n",
    "3. Generate training instances that;\n",
    "    1. Include the \"edge\" of the intron with various shifting strategies\n",
    "    2. Exclude the \"edge\" of the intron and get transcribed mRNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| default_exp training.transcription.generation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| hide\n",
    "from nbdev.showdoc import *"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 0. Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "from pathlib import Path\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from pandas.errors import SettingWithCopyWarning\n",
    "import warnings\n",
    "import sqlite3\n",
    "\n",
    "from llm_mito_scanner.data.download import load_config, \\\n",
    "    get_latest_assembly_path, get_genomic_genbank_path\n",
    "from llm_mito_scanner.data.transcription import get_genes\n",
    "from llm_mito_scanner.training.transcription.index import get_intron_locations\n",
    "\n",
    "warnings.filterwarnings(\"ignore\", category=SettingWithCopyWarning)\n",
    "\n",
    "BOS_TOK = \"<bos>\"\n",
    "EOS_TOK = \"<eos>\"\n",
    "INTRON_TOK = \"<intron>\"\n",
    "UNK_TOK = \"<unk>\"\n",
    "NULL_TOK = \"<null>\"\n",
    "PAD_TOK = \"<pad>\"\n",
    "MRNA_BOS_TOK = \"<mrna-bos>\"\n",
    "MRNA_EOS_TOK = \"<mrna-eos>\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| hide\n",
    "config = load_config()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| hide\n",
    "data_path = Path(config.get(\"data_path\"))\n",
    "data_raw_path = data_path / \"raw\"\n",
    "assemblies_path = data_raw_path / \"assemblies\"\n",
    "latest_assembly_path = get_latest_assembly_path(assemblies_path)\n",
    "genomic_genbank_path = get_genomic_genbank_path(latest_assembly_path)\n",
    "gene_db_path = latest_assembly_path / \"genes.db\"\n",
    "training_data_path = latest_assembly_path / \"training\"\n",
    "transcription_data_path = training_data_path / \"transcription\"\n",
    "intron_locations_path = transcription_data_path / \"intron_positions\"\n",
    "for p in [gene_db_path, intron_locations_path]:\n",
    "    if not p.exists():\n",
    "        raise FileNotFoundError(f\"This notebook requires the path {p.resolve()} to exist\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| hide\n",
    "example_chromosome = \"NC_000001.11\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>geneid</th>\n",
       "      <th>sequence</th>\n",
       "      <th>pos_strand</th>\n",
       "      <th>neg_strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>GeneID:10000</td>\n",
       "      <td>ATTGGGCACCGCCCACTTCGTGGGCTTCCAGGTGCGAGCCCTCGCG...</td>\n",
       "      <td>243488232</td>\n",
       "      <td>243851079</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>GeneID:100034743</td>\n",
       "      <td>AAAACCGGAGAGGCGCATCTTGCCGGCTACGGTTGAAACCCGTTCA...</td>\n",
       "      <td>147993861</td>\n",
       "      <td>148014956</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>GeneID:100036575</td>\n",
       "      <td>GTGGGTCTCAGTAGCTCGGGCGGCGGGAGGAGTGGCAGCGGCCAGG...</td>\n",
       "      <td>182941305</td>\n",
       "      <td>182942522</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>GeneID:100101407</td>\n",
       "      <td>AGGCCCCAGGGGGATTTGGCACGGGTACTCCTGTGAGCTGCTGCTC...</td>\n",
       "      <td>45845342</td>\n",
       "      <td>45847719</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>GeneID:100113373</td>\n",
       "      <td>GAAGAAGAGGCGAGAACGACCCCGGGACCGACCAAAGCCCGCGCGC...</td>\n",
       "      <td>8893410</td>\n",
       "      <td>8894603</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     chromosome            geneid   \n",
       "0  NC_000001.11      GeneID:10000  \\\n",
       "1  NC_000001.11  GeneID:100034743   \n",
       "2  NC_000001.11  GeneID:100036575   \n",
       "3  NC_000001.11  GeneID:100101407   \n",
       "4  NC_000001.11  GeneID:100113373   \n",
       "\n",
       "                                            sequence  pos_strand  neg_strand  \n",
       "0  ATTGGGCACCGCCCACTTCGTGGGCTTCCAGGTGCGAGCCCTCGCG...   243488232   243851079  \n",
       "1  AAAACCGGAGAGGCGCATCTTGCCGGCTACGGTTGAAACCCGTTCA...   147993861   148014956  \n",
       "2  GTGGGTCTCAGTAGCTCGGGCGGCGGGAGGAGTGGCAGCGGCCAGG...   182941305   182942522  \n",
       "3  AGGCCCCAGGGGGATTTGGCACGGGTACTCCTGTGAGCTGCTGCTC...    45845342    45847719  \n",
       "4  GAAGAAGAGGCGAGAACGACCCCGGGACCGACCAAAGCCCGCGCGC...     8893410     8894603  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| hide\n",
    "genes = get_genes(latest_assembly_path, chromosome=example_chromosome)\n",
    "genes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>geneid</th>\n",
       "      <th>transcriptid</th>\n",
       "      <th>intron_start</th>\n",
       "      <th>intron_end</th>\n",
       "      <th>mrna_start</th>\n",
       "      <th>mrna_end</th>\n",
       "      <th>chromosome</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GeneID:79501</td>\n",
       "      <td>NM_001005484.2</td>\n",
       "      <td>15</td>\n",
       "      <td>101</td>\n",
       "      <td>0</td>\n",
       "      <td>6167</td>\n",
       "      <td>NC_000001.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GeneID:79501</td>\n",
       "      <td>NM_001005484.2</td>\n",
       "      <td>155</td>\n",
       "      <td>3618</td>\n",
       "      <td>0</td>\n",
       "      <td>6167</td>\n",
       "      <td>NC_000001.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GeneID:112268260</td>\n",
       "      <td>XM_047436352.1</td>\n",
       "      <td>186</td>\n",
       "      <td>547</td>\n",
       "      <td>0</td>\n",
       "      <td>17102</td>\n",
       "      <td>NC_000001.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GeneID:112268260</td>\n",
       "      <td>XM_047436352.1</td>\n",
       "      <td>1339</td>\n",
       "      <td>2365</td>\n",
       "      <td>0</td>\n",
       "      <td>17102</td>\n",
       "      <td>NC_000001.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GeneID:112268260</td>\n",
       "      <td>XM_047436352.1</td>\n",
       "      <td>2467</td>\n",
       "      <td>8912</td>\n",
       "      <td>0</td>\n",
       "      <td>17102</td>\n",
       "      <td>NC_000001.11</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             geneid    transcriptid  intron_start  intron_end  mrna_start   \n",
       "0      GeneID:79501  NM_001005484.2            15         101           0  \\\n",
       "1      GeneID:79501  NM_001005484.2           155        3618           0   \n",
       "2  GeneID:112268260  XM_047436352.1           186         547           0   \n",
       "3  GeneID:112268260  XM_047436352.1          1339        2365           0   \n",
       "4  GeneID:112268260  XM_047436352.1          2467        8912           0   \n",
       "\n",
       "   mrna_end    chromosome  \n",
       "0      6167  NC_000001.11  \n",
       "1      6167  NC_000001.11  \n",
       "2     17102  NC_000001.11  \n",
       "3     17102  NC_000001.11  \n",
       "4     17102  NC_000001.11  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| hide\n",
    "intron_locations = get_intron_locations(intron_locations_path, chromosome=example_chromosome)\n",
    "intron_locations.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>geneid</th>\n",
       "      <th>transcriptid</th>\n",
       "      <th>intron_start</th>\n",
       "      <th>intron_end</th>\n",
       "      <th>mrna_start</th>\n",
       "      <th>mrna_end</th>\n",
       "      <th>chromosome</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>GeneID:81399</td>\n",
       "      <td>XM_017002410.2</td>\n",
       "      <td>82</td>\n",
       "      <td>22713</td>\n",
       "      <td>8309</td>\n",
       "      <td>44026</td>\n",
       "      <td>NC_000001.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>GeneID:81399</td>\n",
       "      <td>XM_017002408.2</td>\n",
       "      <td>82</td>\n",
       "      <td>20998</td>\n",
       "      <td>8309</td>\n",
       "      <td>44026</td>\n",
       "      <td>NC_000001.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>GeneID:81399</td>\n",
       "      <td>XM_017002408.2</td>\n",
       "      <td>21216</td>\n",
       "      <td>22713</td>\n",
       "      <td>8309</td>\n",
       "      <td>44026</td>\n",
       "      <td>NC_000001.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>GeneID:81399</td>\n",
       "      <td>XM_047431162.1</td>\n",
       "      <td>19</td>\n",
       "      <td>9866</td>\n",
       "      <td>21156</td>\n",
       "      <td>44026</td>\n",
       "      <td>NC_000001.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>GeneID:148398</td>\n",
       "      <td>NM_152486.4</td>\n",
       "      <td>70</td>\n",
       "      <td>191</td>\n",
       "      <td>1808</td>\n",
       "      <td>20652</td>\n",
       "      <td>NC_000001.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>GeneID:148398</td>\n",
       "      <td>NM_152486.4</td>\n",
       "      <td>283</td>\n",
       "      <td>4424</td>\n",
       "      <td>1808</td>\n",
       "      <td>20652</td>\n",
       "      <td>NC_000001.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>GeneID:148398</td>\n",
       "      <td>NM_152486.4</td>\n",
       "      <td>4606</td>\n",
       "      <td>5308</td>\n",
       "      <td>1808</td>\n",
       "      <td>20652</td>\n",
       "      <td>NC_000001.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>GeneID:148398</td>\n",
       "      <td>NM_152486.4</td>\n",
       "      <td>5359</td>\n",
       "      <td>10041</td>\n",
       "      <td>1808</td>\n",
       "      <td>20652</td>\n",
       "      <td>NC_000001.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>GeneID:148398</td>\n",
       "      <td>NM_152486.4</td>\n",
       "      <td>10166</td>\n",
       "      <td>13309</td>\n",
       "      <td>1808</td>\n",
       "      <td>20652</td>\n",
       "      <td>NC_000001.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>GeneID:148398</td>\n",
       "      <td>NM_152486.4</td>\n",
       "      <td>13399</td>\n",
       "      <td>13544</td>\n",
       "      <td>1808</td>\n",
       "      <td>20652</td>\n",
       "      <td>NC_000001.11</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           geneid    transcriptid  intron_start  intron_end  mrna_start   \n",
       "12   GeneID:81399  XM_017002410.2            82       22713        8309  \\\n",
       "13   GeneID:81399  XM_017002408.2            82       20998        8309   \n",
       "14   GeneID:81399  XM_017002408.2         21216       22713        8309   \n",
       "15   GeneID:81399  XM_047431162.1            19        9866       21156   \n",
       "42  GeneID:148398     NM_152486.4            70         191        1808   \n",
       "43  GeneID:148398     NM_152486.4           283        4424        1808   \n",
       "44  GeneID:148398     NM_152486.4          4606        5308        1808   \n",
       "45  GeneID:148398     NM_152486.4          5359       10041        1808   \n",
       "46  GeneID:148398     NM_152486.4         10166       13309        1808   \n",
       "47  GeneID:148398     NM_152486.4         13399       13544        1808   \n",
       "\n",
       "    mrna_end    chromosome  \n",
       "12     44026  NC_000001.11  \n",
       "13     44026  NC_000001.11  \n",
       "14     44026  NC_000001.11  \n",
       "15     44026  NC_000001.11  \n",
       "42     20652  NC_000001.11  \n",
       "43     20652  NC_000001.11  \n",
       "44     20652  NC_000001.11  \n",
       "45     20652  NC_000001.11  \n",
       "46     20652  NC_000001.11  \n",
       "47     20652  NC_000001.11  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| hide\n",
    "intron_locations[intron_locations.mrna_start != 0].head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>geneid</th>\n",
       "      <th>transcriptid</th>\n",
       "      <th>intron_start</th>\n",
       "      <th>intron_end</th>\n",
       "      <th>mrna_start</th>\n",
       "      <th>mrna_end</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>prev_intron_end</th>\n",
       "      <th>intron_space</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>166725</th>\n",
       "      <td>GeneID:10000</td>\n",
       "      <td>XM_047415643.1</td>\n",
       "      <td>305471</td>\n",
       "      <td>324663</td>\n",
       "      <td>836</td>\n",
       "      <td>325632</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>305469.0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129489</th>\n",
       "      <td>GeneID:9910</td>\n",
       "      <td>XM_047436058.1</td>\n",
       "      <td>680</td>\n",
       "      <td>17669</td>\n",
       "      <td>634643</td>\n",
       "      <td>835789</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>677.0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>131438</th>\n",
       "      <td>GeneID:6646</td>\n",
       "      <td>XM_047428828.1</td>\n",
       "      <td>2441</td>\n",
       "      <td>8430</td>\n",
       "      <td>21210</td>\n",
       "      <td>64884</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>2437.0</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141102</th>\n",
       "      <td>GeneID:7135</td>\n",
       "      <td>NM_003281.4</td>\n",
       "      <td>4615</td>\n",
       "      <td>6476</td>\n",
       "      <td>0</td>\n",
       "      <td>17947</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>4611.0</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>131540</th>\n",
       "      <td>GeneID:126859</td>\n",
       "      <td>XM_011509167.4</td>\n",
       "      <td>167005</td>\n",
       "      <td>167824</td>\n",
       "      <td>266</td>\n",
       "      <td>189031</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>166999.0</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               geneid    transcriptid  intron_start  intron_end  mrna_start   \n",
       "166725   GeneID:10000  XM_047415643.1        305471      324663         836  \\\n",
       "129489    GeneID:9910  XM_047436058.1           680       17669      634643   \n",
       "131438    GeneID:6646  XM_047428828.1          2441        8430       21210   \n",
       "141102    GeneID:7135     NM_003281.4          4615        6476           0   \n",
       "131540  GeneID:126859  XM_011509167.4        167005      167824         266   \n",
       "\n",
       "        mrna_end    chromosome  prev_intron_end  intron_space  \n",
       "166725    325632  NC_000001.11         305469.0             2  \n",
       "129489    835789  NC_000001.11            677.0             3  \n",
       "131438     64884  NC_000001.11           2437.0             4  \n",
       "141102     17947  NC_000001.11           4611.0             4  \n",
       "131540    189031  NC_000001.11         166999.0             6  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| hide\n",
    "prev_intron_end = intron_locations.groupby(['chromosome', 'geneid', 'transcriptid']).intron_end.shift(1)\n",
    "intron_space = pd.concat([intron_locations, prev_intron_end], axis=1).dropna()\n",
    "intron_space.columns = intron_locations.columns.tolist() + ['prev_intron_end']\n",
    "intron_space.loc[:, 'intron_space'] = (intron_space.intron_start - intron_space.prev_intron_end).astype(int)\n",
    "intron_space.sort_values('intron_space', ascending=True, inplace=True)\n",
    "intron_space.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    155915.000000\n",
       "mean        153.005388\n",
       "std         198.402928\n",
       "min           2.000000\n",
       "25%          89.000000\n",
       "50%         125.000000\n",
       "75%         171.000000\n",
       "max       14864.000000\n",
       "Name: intron_space, dtype: float64"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| hide\n",
    "intron_space.intron_space.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('NC_000001.11', 'GeneID:148398')"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| hide\n",
    "example_intron = intron_locations.iloc[42, :]\n",
    "example_chromosome = example_intron.chromosome\n",
    "example_gene = example_intron.geneid\n",
    "example_chromosome, example_gene"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. mRNA Sequence Generation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def get_gene(assembly_path: Path, geneid: str) -> str:\n",
    "    genes = get_genes(assembly_path, gene_ids=[geneid])\n",
    "    return genes.set_index('geneid').sequence.iloc[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "20652"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| hide\n",
    "example_gene_sequence = get_gene(latest_assembly_path, example_gene)\n",
    "len(example_gene_sequence)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def get_mrna_intron_locations(\n",
    "        chromosome: str, gene_id: str, transcript_id: str, \n",
    "        intron_locations: pd.DataFrame = None, intron_locations_path: Path = None\n",
    "        ) -> list[tuple[int, int]]:\n",
    "    if intron_locations is None:\n",
    "        intron_locations = get_intron_locations(intron_locations_path)\n",
    "    mrna_intron_locations = intron_locations[\n",
    "        (intron_locations.chromosome == chromosome) &\n",
    "        (intron_locations.geneid == gene_id) &\n",
    "        (intron_locations.transcriptid == transcript_id)\n",
    "    ]\n",
    "    if mrna_intron_locations.shape[0] == 0:\n",
    "        return []\n",
    "    mrna_intron_locations_list = list(map(tuple, mrna_intron_locations[['intron_start', 'intron_end']].values.tolist()))\n",
    "    return mrna_intron_locations_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "13\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[(70, 191),\n",
       " (283, 4424),\n",
       " (4606, 5308),\n",
       " (5359, 10041),\n",
       " (10166, 13309),\n",
       " (13399, 13544),\n",
       " (13730, 15413),\n",
       " (15576, 16405),\n",
       " (16521, 16679),\n",
       " (16758, 16828),\n",
       " (17328, 17522),\n",
       " (17647, 17967),\n",
       " (18078, 18177)]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| hide\n",
    "example_mrna_intron_locations = get_mrna_intron_locations(\n",
    "    example_chromosome, example_gene, example_intron.transcriptid,\n",
    "    intron_locations=intron_locations\n",
    ")\n",
    "print(len(example_mrna_intron_locations))\n",
    "example_mrna_intron_locations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def get_mrna_from_gene(\n",
    "        gene_sequence: str, \n",
    "        mrna_start: int, mrna_end: int, \n",
    "        intron_locations: list[tuple[int, int]],\n",
    "        intron_token: str = INTRON_TOK,\n",
    "        untranscribed_token: str = NULL_TOK,\n",
    "        sep: str = \",\",\n",
    "        debug: bool = False) -> str:\n",
    "    \"Get annotated input and target sequences for a given mRNA.\"\n",
    "    gene_sequence_length = len(gene_sequence)\n",
    "    start_pad_len = mrna_start\n",
    "    end_pad_len = gene_sequence_length - mrna_end\n",
    "    if debug:\n",
    "        print(\"MRNA START:\\t\", mrna_start)\n",
    "        print(\"MRNA END:\\t\", mrna_end)\n",
    "        print(\"GENE SEQUENCE LENGTH:\\t\", gene_sequence_length)\n",
    "        print(\"START PAD LENGTH:\\t\", start_pad_len)\n",
    "        print(\"END PAD LENGTH:\\t\", end_pad_len)\n",
    "    mrna = list(gene_sequence[mrna_start: mrna_end + 1])\n",
    "    for intron_start, intron_end in intron_locations:\n",
    "        pre_intron = mrna[:intron_start]\n",
    "        post_intron = mrna[intron_end:]\n",
    "        intron_length = intron_end - intron_start\n",
    "        intron = [intron_token] * intron_length\n",
    "        mrna = pre_intron + intron + post_intron\n",
    "    if start_pad_len > 0:\n",
    "        start_pad = ([untranscribed_token] * start_pad_len)\n",
    "        mrna = start_pad + mrna\n",
    "    if end_pad_len > 0:\n",
    "        end_pad = ([untranscribed_token] * end_pad_len)\n",
    "        mrna = mrna + end_pad\n",
    "    return sep.join([n if n != \"T\" else \"U\" for n in mrna])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>geneid</th>\n",
       "      <th>transcriptid</th>\n",
       "      <th>mrna_start</th>\n",
       "      <th>mrna_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>GeneID:79501</td>\n",
       "      <td>NM_001005484.2</td>\n",
       "      <td>0</td>\n",
       "      <td>6167</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>GeneID:112268260</td>\n",
       "      <td>XM_047436352.1</td>\n",
       "      <td>0</td>\n",
       "      <td>17102</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>GeneID:105378947</td>\n",
       "      <td>XM_011542538.1</td>\n",
       "      <td>0</td>\n",
       "      <td>25011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>GeneID:81399</td>\n",
       "      <td>XM_024449992.2</td>\n",
       "      <td>0</td>\n",
       "      <td>44026</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>GeneID:81399</td>\n",
       "      <td>XM_017002410.2</td>\n",
       "      <td>8309</td>\n",
       "      <td>44026</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>GeneID:81399</td>\n",
       "      <td>XM_017002408.2</td>\n",
       "      <td>8309</td>\n",
       "      <td>44026</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>GeneID:81399</td>\n",
       "      <td>XM_047431162.1</td>\n",
       "      <td>21156</td>\n",
       "      <td>44026</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>GeneID:148398</td>\n",
       "      <td>NM_001385640.1</td>\n",
       "      <td>0</td>\n",
       "      <td>20652</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>GeneID:148398</td>\n",
       "      <td>NM_001385641.1</td>\n",
       "      <td>0</td>\n",
       "      <td>20652</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>GeneID:148398</td>\n",
       "      <td>NM_152486.4</td>\n",
       "      <td>1808</td>\n",
       "      <td>20652</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      chromosome            geneid    transcriptid  mrna_start  mrna_end\n",
       "0   NC_000001.11      GeneID:79501  NM_001005484.2           0      6167\n",
       "2   NC_000001.11  GeneID:112268260  XM_047436352.1           0     17102\n",
       "6   NC_000001.11  GeneID:105378947  XM_011542538.1           0     25011\n",
       "11  NC_000001.11      GeneID:81399  XM_024449992.2           0     44026\n",
       "12  NC_000001.11      GeneID:81399  XM_017002410.2        8309     44026\n",
       "13  NC_000001.11      GeneID:81399  XM_017002408.2        8309     44026\n",
       "15  NC_000001.11      GeneID:81399  XM_047431162.1       21156     44026\n",
       "16  NC_000001.11     GeneID:148398  NM_001385640.1           0     20652\n",
       "29  NC_000001.11     GeneID:148398  NM_001385641.1           0     20652\n",
       "42  NC_000001.11     GeneID:148398     NM_152486.4        1808     20652"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| hide\n",
    "example_mrna = intron_locations[['chromosome', 'geneid', 'transcriptid', 'mrna_start', 'mrna_end']].drop_duplicates().head(10)\n",
    "example_mrna"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| hide\n",
    "example_mrna_sequences = example_mrna.apply(lambda row: get_mrna_from_gene(\n",
    "    genes[(genes.geneid == row.geneid)].iloc[0, :].sequence,\n",
    "    row.mrna_start, row.mrna_end,\n",
    "    get_mrna_intron_locations(\n",
    "        row.chromosome, row.geneid, row.transcriptid, \n",
    "        intron_locations[intron_locations.chromosome==row.chromosome])), \n",
    "    axis=1)\n",
    "example_mrna_sequences.name = \"sequence\"\n",
    "example_mrna_sequences = pd.concat(\n",
    "    [\n",
    "        example_mrna,\n",
    "        example_mrna_sequences\n",
    "    ],axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>geneid</th>\n",
       "      <th>transcriptid</th>\n",
       "      <th>mrna_start</th>\n",
       "      <th>mrna_end</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>GeneID:79501</td>\n",
       "      <td>NM_001005484.2</td>\n",
       "      <td>0</td>\n",
       "      <td>6167</td>\n",
       "      <td>C,C,C,A,G,A,U,C,U,C,U,U,C,A,G,&lt;intron&gt;,&lt;intron...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>GeneID:112268260</td>\n",
       "      <td>XM_047436352.1</td>\n",
       "      <td>0</td>\n",
       "      <td>17102</td>\n",
       "      <td>A,U,G,C,C,U,A,G,A,C,A,C,A,C,A,C,A,U,C,C,U,U,A,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>GeneID:105378947</td>\n",
       "      <td>XM_011542538.1</td>\n",
       "      <td>0</td>\n",
       "      <td>25011</td>\n",
       "      <td>A,U,G,C,G,U,A,G,A,C,A,C,A,C,A,C,A,U,C,C,U,U,A,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>GeneID:81399</td>\n",
       "      <td>XM_024449992.2</td>\n",
       "      <td>0</td>\n",
       "      <td>44026</td>\n",
       "      <td>U,A,U,A,A,A,A,U,G,A,A,A,G,C,U,G,C,C,U,C,U,G,A,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>GeneID:81399</td>\n",
       "      <td>XM_017002410.2</td>\n",
       "      <td>8309</td>\n",
       "      <td>44026</td>\n",
       "      <td>&lt;null&gt;,&lt;null&gt;,&lt;null&gt;,&lt;null&gt;,&lt;null&gt;,&lt;null&gt;,&lt;nul...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>GeneID:81399</td>\n",
       "      <td>XM_017002408.2</td>\n",
       "      <td>8309</td>\n",
       "      <td>44026</td>\n",
       "      <td>&lt;null&gt;,&lt;null&gt;,&lt;null&gt;,&lt;null&gt;,&lt;null&gt;,&lt;null&gt;,&lt;nul...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>GeneID:81399</td>\n",
       "      <td>XM_047431162.1</td>\n",
       "      <td>21156</td>\n",
       "      <td>44026</td>\n",
       "      <td>&lt;null&gt;,&lt;null&gt;,&lt;null&gt;,&lt;null&gt;,&lt;null&gt;,&lt;null&gt;,&lt;nul...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>GeneID:148398</td>\n",
       "      <td>NM_001385640.1</td>\n",
       "      <td>0</td>\n",
       "      <td>20652</td>\n",
       "      <td>G,G,C,G,G,C,G,G,A,G,U,C,U,C,C,C,A,A,G,U,C,C,C,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>GeneID:148398</td>\n",
       "      <td>NM_001385641.1</td>\n",
       "      <td>0</td>\n",
       "      <td>20652</td>\n",
       "      <td>G,G,C,G,G,C,G,G,A,G,U,C,U,C,C,C,A,A,G,U,C,C,C,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>GeneID:148398</td>\n",
       "      <td>NM_152486.4</td>\n",
       "      <td>1808</td>\n",
       "      <td>20652</td>\n",
       "      <td>&lt;null&gt;,&lt;null&gt;,&lt;null&gt;,&lt;null&gt;,&lt;null&gt;,&lt;null&gt;,&lt;nul...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      chromosome            geneid    transcriptid  mrna_start  mrna_end   \n",
       "0   NC_000001.11      GeneID:79501  NM_001005484.2           0      6167  \\\n",
       "2   NC_000001.11  GeneID:112268260  XM_047436352.1           0     17102   \n",
       "6   NC_000001.11  GeneID:105378947  XM_011542538.1           0     25011   \n",
       "11  NC_000001.11      GeneID:81399  XM_024449992.2           0     44026   \n",
       "12  NC_000001.11      GeneID:81399  XM_017002410.2        8309     44026   \n",
       "13  NC_000001.11      GeneID:81399  XM_017002408.2        8309     44026   \n",
       "15  NC_000001.11      GeneID:81399  XM_047431162.1       21156     44026   \n",
       "16  NC_000001.11     GeneID:148398  NM_001385640.1           0     20652   \n",
       "29  NC_000001.11     GeneID:148398  NM_001385641.1           0     20652   \n",
       "42  NC_000001.11     GeneID:148398     NM_152486.4        1808     20652   \n",
       "\n",
       "                                             sequence  \n",
       "0   C,C,C,A,G,A,U,C,U,C,U,U,C,A,G,<intron>,<intron...  \n",
       "2   A,U,G,C,C,U,A,G,A,C,A,C,A,C,A,C,A,U,C,C,U,U,A,...  \n",
       "6   A,U,G,C,G,U,A,G,A,C,A,C,A,C,A,C,A,U,C,C,U,U,A,...  \n",
       "11  U,A,U,A,A,A,A,U,G,A,A,A,G,C,U,G,C,C,U,C,U,G,A,...  \n",
       "12  <null>,<null>,<null>,<null>,<null>,<null>,<nul...  \n",
       "13  <null>,<null>,<null>,<null>,<null>,<null>,<nul...  \n",
       "15  <null>,<null>,<null>,<null>,<null>,<null>,<nul...  \n",
       "16  G,G,C,G,G,C,G,G,A,G,U,C,U,C,C,C,A,A,G,U,C,C,C,...  \n",
       "29  G,G,C,G,G,C,G,G,A,G,U,C,U,C,C,C,A,A,G,U,C,C,C,...  \n",
       "42  <null>,<null>,<null>,<null>,<null>,<null>,<nul...  "
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| hide\n",
    "example_mrna_sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MRNA START:\t 1808\n",
      "MRNA END:\t 20652\n",
      "GENE SEQUENCE LENGTH:\t 20652\n",
      "START PAD LENGTH:\t 1808\n",
      "END PAD LENGTH:\t 0\n",
      "18844\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "20652"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| hide\n",
    "example_annotated_mrna = get_mrna_from_gene(\n",
    "    example_gene_sequence,\n",
    "    example_intron.mrna_start, example_intron.mrna_end,\n",
    "    example_mrna_intron_locations,\n",
    "    debug=True\n",
    ")\n",
    "print(example_intron.mrna_end - example_intron.mrna_start)\n",
    "len(example_annotated_mrna.split(\",\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(str,\n",
       " 0,\n",
       " 385,\n",
       " 16287,\n",
       " ['<null>', '<null>', '<null>', '<null>', '<null>'],\n",
       " ['G', 'C', 'C', 'U', 'G'])"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| hide\n",
    "type(example_annotated_mrna), example_annotated_mrna.count(\"T\"), example_annotated_mrna.count(\"U\"), \\\n",
    "    example_annotated_mrna.count(\"<intron>\"), example_annotated_mrna.split(\",\")[0:5], example_annotated_mrna.split(\",\")[-5:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def format_mrna_insert_values(\n",
    "        chromosome: str, gene_id: str, transcript_id: str, \n",
    "        sequence: str, start: int, end: int) -> str:\n",
    "    return f\"('{chromosome}', '{gene_id}', '{transcript_id}', '{sequence}', {start}, {end})\"\n",
    "\n",
    "\n",
    "def write_mrna(assembly_path: Path, chromosome: str, mrna: pd.DataFrame, chunk_size: int = 1000):\n",
    "    mrna_db_path = assembly_path / \"mrna.db\"\n",
    "    con = sqlite3.connect(mrna_db_path)\n",
    "    cursor = con.cursor()\n",
    "    try:\n",
    "        cursor.execute(\n",
    "            \"\"\"\n",
    "            CREATE TABLE IF NOT EXISTS mrna (\n",
    "                chromosome TEXT, geneid TEXT, transcriptid TEXT, sequence TEXT, start INT, end INT,\n",
    "                PRIMARY KEY (chromosome, geneid, transcriptid));\"\"\")\n",
    "        chunk_num = max(1, int(mrna.shape[0] / chunk_size))\n",
    "        mrna = np.array_split(mrna, chunk_num)\n",
    "        for mrna_chunk in mrna:\n",
    "            mrna_chunk_insert_statements = mrna_chunk.apply(\n",
    "                lambda row: format_mrna_insert_values(\n",
    "                    chromosome,\n",
    "                    row.geneid,\n",
    "                    row.transcriptid,\n",
    "                    row.sequence,\n",
    "                    row.mrna_start,\n",
    "                    row.mrna_end\n",
    "                ), axis=1).tolist()\n",
    "            batch_insert_script = \"\"\"\n",
    "            BEGIN;\n",
    "            INSERT INTO mrna (chromosome, geneid, transcriptid, sequence, start, end) \"\"\" +\\\n",
    "            \"VALUES \" + \",\".join(mrna_chunk_insert_statements) +\\\n",
    "            \" ON CONFLICT(chromosome, geneid, transcriptid) DO NOTHING;\" +\\\n",
    "            \"COMMIT;\"\n",
    "            cursor.executescript(batch_insert_script)\n",
    "    except Exception as e:\n",
    "        raise e\n",
    "    finally:\n",
    "        cursor.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| hide\n",
    "write_mrna(latest_assembly_path, example_chromosome, example_mrna_sequences)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def get_mrna(\n",
    "        assembly_path: Path, \n",
    "        columns: list[str] = [\n",
    "            \"chromosome\", \"geneid\", \"transcriptid\", \"sequence\", \n",
    "            \"start\", \"end\"],\n",
    "        chromosome: str = None, \n",
    "        gene_ids: list[str] = None, \n",
    "        transcript_ids: list[str] = None, \n",
    "        limit: int = None) -> pd.DataFrame:\n",
    "    query = f\"SELECT {','.join(columns)} FROM mrna\"\n",
    "    if isinstance(chromosome, str):\n",
    "        query = query + f\" WHERE chromosome='{chromosome}'\"\n",
    "    if isinstance(gene_ids, list):\n",
    "        gene_ids_str = \",\".join([\"'\" + i + \"'\" for i in gene_ids])\n",
    "        if \"WHERE\" in query:\n",
    "            query = query + f\" AND geneid IN ({gene_ids_str})\"\n",
    "        else:\n",
    "            query = query + f\" WHERE geneid IN ({gene_ids_str})\"\n",
    "    if isinstance(transcript_ids, list):\n",
    "        transcript_ids_str = \",\".join([\"'\" + i + \"'\" for i in transcript_ids])\n",
    "        if \"WHERE\" in query:\n",
    "            query = query + f\" AND transcriptid IN ({transcript_ids_str})\"\n",
    "        else:\n",
    "            query = query + f\" WHERE transcriptid IN ({transcript_ids_str})\"\n",
    "    if isinstance(limit, int):\n",
    "        query = query + f\" LIMIT {limit}\"\n",
    "    try:\n",
    "        con = sqlite3.connect(assembly_path / \"mrna.db\")\n",
    "        mrna = pd.read_sql_query(query, con=con)\n",
    "    except Exception as e:\n",
    "        raise e\n",
    "    finally:\n",
    "        con.close()\n",
    "    return mrna"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>geneid</th>\n",
       "      <th>transcriptid</th>\n",
       "      <th>sequence</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>GeneID:79501</td>\n",
       "      <td>NM_001005484.2</td>\n",
       "      <td>C,C,C,A,G,A,U,C,U,C,U,U,C,A,G,&lt;intron&gt;,&lt;intron...</td>\n",
       "      <td>0</td>\n",
       "      <td>6167</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>GeneID:112268260</td>\n",
       "      <td>XM_047436352.1</td>\n",
       "      <td>A,U,G,C,C,U,A,G,A,C,A,C,A,C,A,C,A,U,C,C,U,U,A,...</td>\n",
       "      <td>0</td>\n",
       "      <td>17102</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>GeneID:105378947</td>\n",
       "      <td>XM_011542538.1</td>\n",
       "      <td>A,U,G,C,G,U,A,G,A,C,A,C,A,C,A,C,A,U,C,C,U,U,A,...</td>\n",
       "      <td>0</td>\n",
       "      <td>25011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>GeneID:81399</td>\n",
       "      <td>XM_024449992.2</td>\n",
       "      <td>U,A,U,A,A,A,A,U,G,A,A,A,G,C,U,G,C,C,U,C,U,G,A,...</td>\n",
       "      <td>0</td>\n",
       "      <td>44026</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>GeneID:81399</td>\n",
       "      <td>XM_017002410.2</td>\n",
       "      <td>&lt;null&gt;,&lt;null&gt;,&lt;null&gt;,&lt;null&gt;,&lt;null&gt;,&lt;null&gt;,&lt;nul...</td>\n",
       "      <td>8309</td>\n",
       "      <td>44026</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     chromosome            geneid    transcriptid   \n",
       "0  NC_000001.11      GeneID:79501  NM_001005484.2  \\\n",
       "1  NC_000001.11  GeneID:112268260  XM_047436352.1   \n",
       "2  NC_000001.11  GeneID:105378947  XM_011542538.1   \n",
       "3  NC_000001.11      GeneID:81399  XM_024449992.2   \n",
       "4  NC_000001.11      GeneID:81399  XM_017002410.2   \n",
       "\n",
       "                                            sequence  start    end  \n",
       "0  C,C,C,A,G,A,U,C,U,C,U,U,C,A,G,<intron>,<intron...      0   6167  \n",
       "1  A,U,G,C,C,U,A,G,A,C,A,C,A,C,A,C,A,U,C,C,U,U,A,...      0  17102  \n",
       "2  A,U,G,C,G,U,A,G,A,C,A,C,A,C,A,C,A,U,C,C,U,U,A,...      0  25011  \n",
       "3  U,A,U,A,A,A,A,U,G,A,A,A,G,C,U,G,C,C,U,C,U,G,A,...      0  44026  \n",
       "4  <null>,<null>,<null>,<null>,<null>,<null>,<nul...   8309  44026  "
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| hide\n",
    "example_mrna_df = get_mrna(latest_assembly_path, limit = 5)\n",
    "example_mrna_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def get_mrna_locations(locations: pd.DataFrame) -> pd.DataFrame:\n",
    "    \"Get the mrna sequences between introns\"\n",
    "    # Get locations of transcribed dna\n",
    "    locations = locations.copy()\n",
    "    locations_grouped = locations.groupby(\n",
    "            ['chromosome', 'geneid', 'transcriptid']\n",
    "    )\n",
    "    mrna_seq_start = (\n",
    "        locations_grouped.intron_end.shift(1).fillna(0)\n",
    "    ).astype(int)\n",
    "    mrna_seq_start.name = \"start\"\n",
    "    mrna_seq_end = locations.intron_start\n",
    "    mrna_seq_end.name = \"end\"\n",
    "    selected_cols = ['chromosome', 'geneid', 'transcriptid', 'mrna_start', 'mrna_end']\n",
    "    mrna_sequences = pd.concat(\n",
    "        [\n",
    "            locations[selected_cols],\n",
    "            mrna_seq_start,\n",
    "            mrna_seq_end\n",
    "        ], axis=1)\n",
    "    # Add last sequence of mRNA from last intron\n",
    "    last_intron = locations_grouped.tail(1)\n",
    "    last_intron.loc[:, 'start'] = last_intron.intron_end\n",
    "    last_intron.loc[:, 'end'] = last_intron.mrna_end\n",
    "    all_mrna_sequences = pd.concat(\n",
    "        [\n",
    "            mrna_sequences,\n",
    "            last_intron[selected_cols + ['start', 'end']]\n",
    "        ], axis=0, ignore_index=True\n",
    "    ).sort_values(['chromosome', 'geneid', 'transcriptid', 'start']).reset_index(drop=True)\n",
    "    return all_mrna_sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>geneid</th>\n",
       "      <th>transcriptid</th>\n",
       "      <th>mrna_start</th>\n",
       "      <th>mrna_end</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>GeneID:10000</td>\n",
       "      <td>NM_001206729.2</td>\n",
       "      <td>7495</td>\n",
       "      <td>351356</td>\n",
       "      <td>0</td>\n",
       "      <td>41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>GeneID:10000</td>\n",
       "      <td>NM_001206729.2</td>\n",
       "      <td>7495</td>\n",
       "      <td>351356</td>\n",
       "      <td>302</td>\n",
       "      <td>460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>GeneID:10000</td>\n",
       "      <td>NM_001206729.2</td>\n",
       "      <td>7495</td>\n",
       "      <td>351356</td>\n",
       "      <td>147868</td>\n",
       "      <td>147994</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>GeneID:10000</td>\n",
       "      <td>NM_001206729.2</td>\n",
       "      <td>7495</td>\n",
       "      <td>351356</td>\n",
       "      <td>178701</td>\n",
       "      <td>178813</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>GeneID:10000</td>\n",
       "      <td>NM_001206729.2</td>\n",
       "      <td>7495</td>\n",
       "      <td>351356</td>\n",
       "      <td>197547</td>\n",
       "      <td>197692</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     chromosome        geneid    transcriptid  mrna_start  mrna_end   start   \n",
       "0  NC_000001.11  GeneID:10000  NM_001206729.2        7495    351356       0  \\\n",
       "1  NC_000001.11  GeneID:10000  NM_001206729.2        7495    351356     302   \n",
       "2  NC_000001.11  GeneID:10000  NM_001206729.2        7495    351356  147868   \n",
       "3  NC_000001.11  GeneID:10000  NM_001206729.2        7495    351356  178701   \n",
       "4  NC_000001.11  GeneID:10000  NM_001206729.2        7495    351356  197547   \n",
       "\n",
       "      end  \n",
       "0      41  \n",
       "1     460  \n",
       "2  147994  \n",
       "3  178813  \n",
       "4  197692  "
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| hide\n",
    "mrna_locations = get_mrna_locations(intron_locations)\n",
    "mrna_locations.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(181355, 7)"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| hide\n",
    "mrna_locations.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>geneid</th>\n",
       "      <th>transcriptid</th>\n",
       "      <th>intron_start</th>\n",
       "      <th>intron_end</th>\n",
       "      <th>mrna_start</th>\n",
       "      <th>mrna_end</th>\n",
       "      <th>chromosome</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>166694</th>\n",
       "      <td>GeneID:10000</td>\n",
       "      <td>NM_001206729.2</td>\n",
       "      <td>41</td>\n",
       "      <td>302</td>\n",
       "      <td>7495</td>\n",
       "      <td>351356</td>\n",
       "      <td>NC_000001.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166695</th>\n",
       "      <td>GeneID:10000</td>\n",
       "      <td>NM_001206729.2</td>\n",
       "      <td>460</td>\n",
       "      <td>147868</td>\n",
       "      <td>7495</td>\n",
       "      <td>351356</td>\n",
       "      <td>NC_000001.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166696</th>\n",
       "      <td>GeneID:10000</td>\n",
       "      <td>NM_001206729.2</td>\n",
       "      <td>147994</td>\n",
       "      <td>178701</td>\n",
       "      <td>7495</td>\n",
       "      <td>351356</td>\n",
       "      <td>NC_000001.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166697</th>\n",
       "      <td>GeneID:10000</td>\n",
       "      <td>NM_001206729.2</td>\n",
       "      <td>178813</td>\n",
       "      <td>197547</td>\n",
       "      <td>7495</td>\n",
       "      <td>351356</td>\n",
       "      <td>NC_000001.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166698</th>\n",
       "      <td>GeneID:10000</td>\n",
       "      <td>NM_001206729.2</td>\n",
       "      <td>197692</td>\n",
       "      <td>205842</td>\n",
       "      <td>7495</td>\n",
       "      <td>351356</td>\n",
       "      <td>NC_000001.11</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              geneid    transcriptid  intron_start  intron_end  mrna_start   \n",
       "166694  GeneID:10000  NM_001206729.2            41         302        7495  \\\n",
       "166695  GeneID:10000  NM_001206729.2           460      147868        7495   \n",
       "166696  GeneID:10000  NM_001206729.2        147994      178701        7495   \n",
       "166697  GeneID:10000  NM_001206729.2        178813      197547        7495   \n",
       "166698  GeneID:10000  NM_001206729.2        197692      205842        7495   \n",
       "\n",
       "        mrna_end    chromosome  \n",
       "166694    351356  NC_000001.11  \n",
       "166695    351356  NC_000001.11  \n",
       "166696    351356  NC_000001.11  \n",
       "166697    351356  NC_000001.11  \n",
       "166698    351356  NC_000001.11  "
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| hide\n",
    "first_sample_mrna = mrna_locations.iloc[0, :]\n",
    "intron_locations[\n",
    "    (intron_locations.chromosome == first_sample_mrna.chromosome) &\n",
    "    (intron_locations.geneid == first_sample_mrna.geneid) &\n",
    "    (intron_locations.transcriptid == first_sample_mrna.transcriptid)\n",
    "].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| hide\n",
    "import nbdev; nbdev.nbdev_export()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
