{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Generating training instances from genomic dna, intron and mRNA positions.\n",
    "\n",
    "> Preparing training instances."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Goals\n",
    "1. Construct the mRNA sequence\n",
    "3. Generate training instances that;\n",
    "    1. Include the \"edge\" of the intron with various shifting strategies\n",
    "    2. Exclude the \"edge\" of the intron and get transcribed mRNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| default_exp training.transcription.generation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 0. Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jdb/projects/llm-mito-scanner/venv/lib/python3.10/site-packages/Bio/__init__.py:138: BiopythonWarning: You may be importing Biopython from inside the source tree. This is bad practice and might lead to downstream issues. In particular, you might encounter ImportErrors due to missing compiled C extensions. We recommend that you try running your code from outside the source tree. If you are outside the source tree then you have a setup.py file in an unexpected directory: /home/jdb/projects/llm-mito-scanner/venv/lib/python3.10/site-packages\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "#| export\n",
    "from pathlib import Path\n",
    "import pandas as pd\n",
    "from pandas.errors import SettingWithCopyWarning\n",
    "import random\n",
    "import warnings\n",
    "import time\n",
    "from datetime import timedelta\n",
    "\n",
    "from llm_mito_scanner.data.download import load_config, \\\n",
    "    get_latest_assembly_path, get_genomic_genbank_path\n",
    "from llm_mito_scanner.data.transcription import read_all_chromosome_gene_info\n",
    "from llm_mito_scanner.training.transcription.index import get_intron_locations\n",
    "\n",
    "warnings.filterwarnings(\"ignore\", category=SettingWithCopyWarning)\n",
    "\n",
    "BOS_TOK = \"<bos>\"\n",
    "EOS_TOK = \"<eos>\"\n",
    "INTRON_TOK = \"<intron>\"\n",
    "UNK_TOK = \"<unk>\"\n",
    "NULL_TOK = \"<null>\"\n",
    "PAD_TOK = \"<pad>\"\n",
    "MRNA_BOS_TOK = \"<mrna-bos>\"\n",
    "MRNA_EOS_TOK = \"<mrna-eos>\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| hide\n",
    "config = load_config()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| hide\n",
    "data_path = Path(config.get(\"data_path\"))\n",
    "data_raw_path = data_path / \"raw\"\n",
    "assemblies_path = data_raw_path / \"assemblies\"\n",
    "latest_assembly_path = get_latest_assembly_path(assemblies_path)\n",
    "genomic_genbank_path = get_genomic_genbank_path(latest_assembly_path)\n",
    "genes_path = latest_assembly_path / \"genes\"\n",
    "training_data_path = latest_assembly_path / \"training\"\n",
    "transcription_data_path = training_data_path / \"transcription\"\n",
    "intron_locations_path = transcription_data_path / \"intron_positions\"\n",
    "for p in [genes_path, intron_locations_path]:\n",
    "    if not p.exists():\n",
    "        raise FileNotFoundError(f\"This notebook requires the path {p.resolve()} to exist\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>geneid</th>\n",
       "      <th>sequence</th>\n",
       "      <th>pos_strand_position</th>\n",
       "      <th>neg_strand_position</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GeneID:100287102</td>\n",
       "      <td>CTTGCCGTCAGCCTTTTCTTTGACCTCTTCTTTCTGTTCATGTGTA...</td>\n",
       "      <td>11873</td>\n",
       "      <td>14409</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GeneID:653635</td>\n",
       "      <td>TCCGGCAGAGCGGAAGCGGCGGCGGGAGCTTCCGGGAGGGCGGCTC...</td>\n",
       "      <td>14361</td>\n",
       "      <td>29370</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GeneID:102466751</td>\n",
       "      <td>TGTGGGAGAGGAACATGGGCTCAGGACAGCGGGTGTCAGCTTGCCT...</td>\n",
       "      <td>17368</td>\n",
       "      <td>17436</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GeneID:107985730</td>\n",
       "      <td>TGCCCTCCAGCCCTACGCCTTGACCCGCTTTCCTGCGTCTCTCAGC...</td>\n",
       "      <td>29773</td>\n",
       "      <td>35418</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GeneID:100302278</td>\n",
       "      <td>GGATGCCCAGCTAGTTTGAATTTTAGATAAACAACGAATAATTTCG...</td>\n",
       "      <td>30365</td>\n",
       "      <td>30503</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             geneid                                           sequence   \n",
       "0  GeneID:100287102  CTTGCCGTCAGCCTTTTCTTTGACCTCTTCTTTCTGTTCATGTGTA...  \\\n",
       "1     GeneID:653635  TCCGGCAGAGCGGAAGCGGCGGCGGGAGCTTCCGGGAGGGCGGCTC...   \n",
       "2  GeneID:102466751  TGTGGGAGAGGAACATGGGCTCAGGACAGCGGGTGTCAGCTTGCCT...   \n",
       "3  GeneID:107985730  TGCCCTCCAGCCCTACGCCTTGACCCGCTTTCCTGCGTCTCTCAGC...   \n",
       "4  GeneID:100302278  GGATGCCCAGCTAGTTTGAATTTTAGATAAACAACGAATAATTTCG...   \n",
       "\n",
       "   pos_strand_position  neg_strand_position  \n",
       "0                11873                14409  \n",
       "1                14361                29370  \n",
       "2                17368                17436  \n",
       "3                29773                35418  \n",
       "4                30365                30503  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| hide\n",
    "genes = read_all_chromosome_gene_info(latest_assembly_path)\n",
    "genes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>geneid</th>\n",
       "      <th>transcriptid</th>\n",
       "      <th>intron_start</th>\n",
       "      <th>intron_end</th>\n",
       "      <th>mrna_start</th>\n",
       "      <th>mrna_end</th>\n",
       "      <th>chromosome</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GeneID:79501</td>\n",
       "      <td>NM_001005484.2</td>\n",
       "      <td>15</td>\n",
       "      <td>101</td>\n",
       "      <td>0</td>\n",
       "      <td>6167</td>\n",
       "      <td>NC_000001.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GeneID:79501</td>\n",
       "      <td>NM_001005484.2</td>\n",
       "      <td>155</td>\n",
       "      <td>3618</td>\n",
       "      <td>0</td>\n",
       "      <td>6167</td>\n",
       "      <td>NC_000001.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GeneID:112268260</td>\n",
       "      <td>XM_047436352.1</td>\n",
       "      <td>186</td>\n",
       "      <td>547</td>\n",
       "      <td>0</td>\n",
       "      <td>17102</td>\n",
       "      <td>NC_000001.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GeneID:112268260</td>\n",
       "      <td>XM_047436352.1</td>\n",
       "      <td>1339</td>\n",
       "      <td>2365</td>\n",
       "      <td>0</td>\n",
       "      <td>17102</td>\n",
       "      <td>NC_000001.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GeneID:112268260</td>\n",
       "      <td>XM_047436352.1</td>\n",
       "      <td>2467</td>\n",
       "      <td>8912</td>\n",
       "      <td>0</td>\n",
       "      <td>17102</td>\n",
       "      <td>NC_000001.11</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             geneid    transcriptid  intron_start  intron_end  mrna_start   \n",
       "0      GeneID:79501  NM_001005484.2            15         101           0  \\\n",
       "1      GeneID:79501  NM_001005484.2           155        3618           0   \n",
       "2  GeneID:112268260  XM_047436352.1           186         547           0   \n",
       "3  GeneID:112268260  XM_047436352.1          1339        2365           0   \n",
       "4  GeneID:112268260  XM_047436352.1          2467        8912           0   \n",
       "\n",
       "   mrna_end    chromosome  \n",
       "0      6167  NC_000001.11  \n",
       "1      6167  NC_000001.11  \n",
       "2     17102  NC_000001.11  \n",
       "3     17102  NC_000001.11  \n",
       "4     17102  NC_000001.11  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| hide\n",
    "intron_locations = get_intron_locations(intron_locations_path)\n",
    "intron_locations.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>geneid</th>\n",
       "      <th>transcriptid</th>\n",
       "      <th>intron_start</th>\n",
       "      <th>intron_end</th>\n",
       "      <th>mrna_start</th>\n",
       "      <th>mrna_end</th>\n",
       "      <th>chromosome</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>GeneID:81399</td>\n",
       "      <td>XM_017002410.2</td>\n",
       "      <td>82</td>\n",
       "      <td>22713</td>\n",
       "      <td>8309</td>\n",
       "      <td>44026</td>\n",
       "      <td>NC_000001.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>GeneID:81399</td>\n",
       "      <td>XM_017002408.2</td>\n",
       "      <td>82</td>\n",
       "      <td>20998</td>\n",
       "      <td>8309</td>\n",
       "      <td>44026</td>\n",
       "      <td>NC_000001.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>GeneID:81399</td>\n",
       "      <td>XM_017002408.2</td>\n",
       "      <td>21216</td>\n",
       "      <td>22713</td>\n",
       "      <td>8309</td>\n",
       "      <td>44026</td>\n",
       "      <td>NC_000001.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>GeneID:81399</td>\n",
       "      <td>XM_047431162.1</td>\n",
       "      <td>19</td>\n",
       "      <td>9866</td>\n",
       "      <td>21156</td>\n",
       "      <td>44026</td>\n",
       "      <td>NC_000001.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>GeneID:148398</td>\n",
       "      <td>NM_152486.4</td>\n",
       "      <td>70</td>\n",
       "      <td>191</td>\n",
       "      <td>1808</td>\n",
       "      <td>20652</td>\n",
       "      <td>NC_000001.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>GeneID:148398</td>\n",
       "      <td>NM_152486.4</td>\n",
       "      <td>283</td>\n",
       "      <td>4424</td>\n",
       "      <td>1808</td>\n",
       "      <td>20652</td>\n",
       "      <td>NC_000001.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>GeneID:148398</td>\n",
       "      <td>NM_152486.4</td>\n",
       "      <td>4606</td>\n",
       "      <td>5308</td>\n",
       "      <td>1808</td>\n",
       "      <td>20652</td>\n",
       "      <td>NC_000001.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>GeneID:148398</td>\n",
       "      <td>NM_152486.4</td>\n",
       "      <td>5359</td>\n",
       "      <td>10041</td>\n",
       "      <td>1808</td>\n",
       "      <td>20652</td>\n",
       "      <td>NC_000001.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>GeneID:148398</td>\n",
       "      <td>NM_152486.4</td>\n",
       "      <td>10166</td>\n",
       "      <td>13309</td>\n",
       "      <td>1808</td>\n",
       "      <td>20652</td>\n",
       "      <td>NC_000001.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>GeneID:148398</td>\n",
       "      <td>NM_152486.4</td>\n",
       "      <td>13399</td>\n",
       "      <td>13544</td>\n",
       "      <td>1808</td>\n",
       "      <td>20652</td>\n",
       "      <td>NC_000001.11</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           geneid    transcriptid  intron_start  intron_end  mrna_start   \n",
       "12   GeneID:81399  XM_017002410.2            82       22713        8309  \\\n",
       "13   GeneID:81399  XM_017002408.2            82       20998        8309   \n",
       "14   GeneID:81399  XM_017002408.2         21216       22713        8309   \n",
       "15   GeneID:81399  XM_047431162.1            19        9866       21156   \n",
       "42  GeneID:148398     NM_152486.4            70         191        1808   \n",
       "43  GeneID:148398     NM_152486.4           283        4424        1808   \n",
       "44  GeneID:148398     NM_152486.4          4606        5308        1808   \n",
       "45  GeneID:148398     NM_152486.4          5359       10041        1808   \n",
       "46  GeneID:148398     NM_152486.4         10166       13309        1808   \n",
       "47  GeneID:148398     NM_152486.4         13399       13544        1808   \n",
       "\n",
       "    mrna_end    chromosome  \n",
       "12     44026  NC_000001.11  \n",
       "13     44026  NC_000001.11  \n",
       "14     44026  NC_000001.11  \n",
       "15     44026  NC_000001.11  \n",
       "42     20652  NC_000001.11  \n",
       "43     20652  NC_000001.11  \n",
       "44     20652  NC_000001.11  \n",
       "45     20652  NC_000001.11  \n",
       "46     20652  NC_000001.11  \n",
       "47     20652  NC_000001.11  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| hide\n",
    "intron_locations[intron_locations.mrna_start != 0].head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>geneid</th>\n",
       "      <th>transcriptid</th>\n",
       "      <th>intron_start</th>\n",
       "      <th>intron_end</th>\n",
       "      <th>mrna_start</th>\n",
       "      <th>mrna_end</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>prev_intron_end</th>\n",
       "      <th>intron_space</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1073192</th>\n",
       "      <td>GeneID:3845</td>\n",
       "      <td>XM_047428826.1</td>\n",
       "      <td>38752</td>\n",
       "      <td>41018</td>\n",
       "      <td>0</td>\n",
       "      <td>45684</td>\n",
       "      <td>NC_000012.12</td>\n",
       "      <td>38751.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>581729</th>\n",
       "      <td>GeneID:2870</td>\n",
       "      <td>NM_002082.4</td>\n",
       "      <td>14571</td>\n",
       "      <td>15061</td>\n",
       "      <td>1153</td>\n",
       "      <td>17100</td>\n",
       "      <td>NC_000005.10</td>\n",
       "      <td>14569.0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>610529</th>\n",
       "      <td>GeneID:23506</td>\n",
       "      <td>XM_024446390.2</td>\n",
       "      <td>1127</td>\n",
       "      <td>19730</td>\n",
       "      <td>44237</td>\n",
       "      <td>122218</td>\n",
       "      <td>NC_000006.12</td>\n",
       "      <td>1125.0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1192146</th>\n",
       "      <td>GeneID:55012</td>\n",
       "      <td>XM_024449638.2</td>\n",
       "      <td>1472</td>\n",
       "      <td>5561</td>\n",
       "      <td>0</td>\n",
       "      <td>36827</td>\n",
       "      <td>NC_000014.9</td>\n",
       "      <td>1470.0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1699082</th>\n",
       "      <td>GeneID:55217</td>\n",
       "      <td>XM_047442236.1</td>\n",
       "      <td>3580</td>\n",
       "      <td>11733</td>\n",
       "      <td>76587</td>\n",
       "      <td>123942</td>\n",
       "      <td>NC_000023.11</td>\n",
       "      <td>3578.0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               geneid    transcriptid  intron_start  intron_end  mrna_start   \n",
       "1073192   GeneID:3845  XM_047428826.1         38752       41018           0  \\\n",
       "581729    GeneID:2870     NM_002082.4         14571       15061        1153   \n",
       "610529   GeneID:23506  XM_024446390.2          1127       19730       44237   \n",
       "1192146  GeneID:55012  XM_024449638.2          1472        5561           0   \n",
       "1699082  GeneID:55217  XM_047442236.1          3580       11733       76587   \n",
       "\n",
       "         mrna_end    chromosome  prev_intron_end  intron_space  \n",
       "1073192     45684  NC_000012.12          38751.0             1  \n",
       "581729      17100  NC_000005.10          14569.0             2  \n",
       "610529     122218  NC_000006.12           1125.0             2  \n",
       "1192146     36827   NC_000014.9           1470.0             2  \n",
       "1699082    123942  NC_000023.11           3578.0             2  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| hide\n",
    "prev_intron_end = intron_locations.groupby(['chromosome', 'geneid', 'transcriptid']).intron_end.shift(1)\n",
    "intron_space = pd.concat([intron_locations, prev_intron_end], axis=1).dropna()\n",
    "intron_space.columns = intron_locations.columns.tolist() + ['prev_intron_end']\n",
    "intron_space.loc[:, 'intron_space'] = (intron_space.intron_start - intron_space.prev_intron_end).astype(int)\n",
    "intron_space.sort_values('intron_space', ascending=True, inplace=True)\n",
    "intron_space.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    1.574709e+06\n",
       "mean     1.568139e+02\n",
       "std      3.197923e+02\n",
       "min      1.000000e+00\n",
       "25%      8.800000e+01\n",
       "50%      1.220000e+02\n",
       "75%      1.660000e+02\n",
       "max      8.761600e+04\n",
       "Name: intron_space, dtype: float64"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| hide\n",
    "intron_space.intron_space.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('NC_000001.11', 'GeneID:148398')"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| hide\n",
    "example_intron = intron_locations.iloc[42, :]\n",
    "example_chromosome = example_intron.chromosome\n",
    "example_gene = example_intron.geneid\n",
    "example_chromosome, example_gene"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. mRNA Sequence Generation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def get_gene(genes: pd.DataFrame, geneid: str) -> str:\n",
    "    return genes.set_index('geneid').loc[geneid].sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "20652"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| hide\n",
    "example_gene_sequence = get_gene(genes, example_gene)\n",
    "len(example_gene_sequence)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def get_mrna_intron_locations(\n",
    "        chromosome: str, gene_id: str, transcript_id: str, \n",
    "        intron_locations: pd.DataFrame = None, intron_locations_path: Path = None\n",
    "        ) -> list[tuple[int, int]]:\n",
    "    if intron_locations is None:\n",
    "        intron_locations = get_intron_locations(intron_locations_path)\n",
    "    mrna_intron_locations = intron_locations[\n",
    "        (intron_locations.chromosome == chromosome) &\n",
    "        (intron_locations.geneid == gene_id) &\n",
    "        (intron_locations.transcriptid == transcript_id)\n",
    "    ]\n",
    "    if mrna_intron_locations.shape[0] == 0:\n",
    "        return []\n",
    "    mrna_intron_locations_list = list(map(tuple, mrna_intron_locations[['intron_start', 'intron_end']].values.tolist()))\n",
    "    return mrna_intron_locations_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "13\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[(70, 191),\n",
       " (283, 4424),\n",
       " (4606, 5308),\n",
       " (5359, 10041),\n",
       " (10166, 13309),\n",
       " (13399, 13544),\n",
       " (13730, 15413),\n",
       " (15576, 16405),\n",
       " (16521, 16679),\n",
       " (16758, 16828),\n",
       " (17328, 17522),\n",
       " (17647, 17967),\n",
       " (18078, 18177)]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| hide\n",
    "example_mrna_intron_locations = get_mrna_intron_locations(\n",
    "    example_chromosome, example_gene, example_intron.transcriptid,\n",
    "    intron_locations=intron_locations\n",
    ")\n",
    "print(len(example_mrna_intron_locations))\n",
    "example_mrna_intron_locations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def get_mrna_from_gene(\n",
    "        gene_sequence: str, \n",
    "        mrna_start: int, mrna_end: int, \n",
    "        intron_locations: list[tuple[int, int]],\n",
    "        intron_token: str = INTRON_TOK,\n",
    "        untranscribed_token: str = NULL_TOK,\n",
    "        debug: bool = False) -> tuple[str, str]:\n",
    "    \"Get annotated input and target sequences for a given mRNA.\"\n",
    "    gene_sequence_length = len(gene_sequence)\n",
    "    start_pad_len = mrna_start\n",
    "    end_pad_len = gene_sequence_length - mrna_end\n",
    "    if debug:\n",
    "        print(\"MRNA START:\\t\", mrna_start)\n",
    "        print(\"MRNA END:\\t\", mrna_end)\n",
    "        print(\"GENE SEQUENCE LENGTH:\\t\", gene_sequence_length)\n",
    "        print(\"START PAD LENGTH:\\t\", start_pad_len)\n",
    "        print(\"END PAD LENGTH:\\t\", end_pad_len)\n",
    "    mrna = list(gene_sequence[mrna_start: mrna_end + 1])\n",
    "    for intron_start, intron_end in intron_locations:\n",
    "        pre_intron = mrna[:intron_start]\n",
    "        post_intron = mrna[intron_end:]\n",
    "        intron_length = intron_end - intron_start\n",
    "        intron = [intron_token] * intron_length\n",
    "        mrna = pre_intron + intron + post_intron\n",
    "    if start_pad_len > 0:\n",
    "        start_pad = ([untranscribed_token] * start_pad_len)\n",
    "        mrna = start_pad + mrna\n",
    "    if end_pad_len > 0:\n",
    "        end_pad = ([untranscribed_token] * end_pad_len)\n",
    "        mrna = mrna + end_pad\n",
    "    gene = list(gene_sequence)\n",
    "    mrna = [n if n != \"T\" else \"U\" for n in mrna]\n",
    "    return gene, mrna"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MRNA START:\t 1808\n",
      "MRNA END:\t 20652\n",
      "GENE SEQUENCE LENGTH:\t 20652\n",
      "START PAD LENGTH:\t 1808\n",
      "END PAD LENGTH:\t 0\n",
      "18844\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(20652, 20652)"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| hide\n",
    "example_annotated_gene, example_annotated_mrna = get_mrna_from_gene(\n",
    "    example_gene_sequence,\n",
    "    example_intron.mrna_start, example_intron.mrna_end,\n",
    "    example_mrna_intron_locations,\n",
    "    debug=True\n",
    ")\n",
    "print(example_intron.mrna_end - example_intron.mrna_start)\n",
    "len(example_annotated_gene), len(example_annotated_mrna)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(list, 3539, 0, 0, ['G', 'G', 'C', 'G', 'G'], ['G', 'C', 'C', 'T', 'G'])"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| hide\n",
    "type(example_annotated_gene), example_annotated_gene.count(\"T\"), example_annotated_gene.count(\"U\"), \\\n",
    "    example_annotated_gene.count(\"<intron>\"), example_annotated_gene[0:5], example_annotated_gene[-5:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(list,\n",
       " 0,\n",
       " 385,\n",
       " 16287,\n",
       " ['<null>', '<null>', '<null>', '<null>', '<null>'],\n",
       " ['G', 'C', 'C', 'U', 'G'])"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| hide\n",
    "type(example_annotated_mrna), example_annotated_mrna.count(\"T\"), example_annotated_mrna.count(\"U\"), \\\n",
    "    example_annotated_mrna.count(\"<intron>\"), example_annotated_mrna[0:5], example_annotated_mrna[-5:]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Generate Training Instances\n",
    "\n",
    "When we train, we'll want to;\n",
    "- load training instances easily\n",
    "- ensure training instances are the same length\n",
    "- ensure the training instances are a sample of the dataset\n",
    "- ensure the sample is representative of the dataset\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "129045"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| hide\n",
    "intron_locations.transcriptid.unique().size"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('12,496,356,010', '195,255,562')"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| hide\n",
    "mrna_length_df = intron_locations.drop_duplicates(subset=['chromosome', 'geneid', 'transcriptid'])[['mrna_start', 'mrna_end']]\n",
    "total_dataset_length = (mrna_length_df.mrna_end - mrna_length_df.mrna_start).sum()\n",
    "training_sequence_length = 64\n",
    "num_training_instances = total_dataset_length // training_sequence_length\n",
    "'{:,}'.format(total_dataset_length), '{:,}'.format(num_training_instances)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "36000.0"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| hide\n",
    "training_hours = 4\n",
    "training_minutes = training_hours * 60\n",
    "training_seconds = training_minutes * 60\n",
    "sequences_per_second = 2.5\n",
    "training_instance_target = training_seconds * sequences_per_second\n",
    "training_instance_target"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1.0"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| hide\n",
    "proportion_intron_edge = 0.25\n",
    "proportion_intron = 0.25\n",
    "proportion_mrna = 0.25\n",
    "proportion_mrna_edge = 0.25\n",
    "\n",
    "random_state = 42\n",
    "\n",
    "sum([proportion_intron_edge, proportion_intron, proportion_mrna, proportion_mrna_edge])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9000\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "3407640"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| hide\n",
    "# each intron has a start edge and an end edge - lets make sure we sample both types\n",
    "allowed_intron_edges = int(training_instance_target * proportion_intron_edge)\n",
    "print(allowed_intron_edges)\n",
    "unique_introns_edges = intron_locations.shape[0] * 2 # start and end edge\n",
    "unique_introns_edges"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def sample_intron_edges(\n",
    "        locations: pd.DataFrame, n: int, \n",
    "        random_state: int = 42, offset: int = -32, length: int = 64) -> pd.DataFrame:\n",
    "    \"Get training instances where either the start of end of an intron is in the center of the sequence.\"\n",
    "    start_n = int(n / 2)\n",
    "    end_n = n - start_n\n",
    "    replace = False if start_n > locations.shape[0] else True\n",
    "    starts = locations.sample(start_n, replace=replace, random_state=random_state)\n",
    "    ends = locations.sample(end_n, replace=replace, random_state=random_state)\n",
    "    frames = []\n",
    "    for f, slice_origin in zip([starts, ends], ['intron_start', 'intron_end']):\n",
    "        f_slice_start = (f[slice_origin] - f.mrna_start + offset).apply(lambda val: max(0, val))\n",
    "        f.loc[:, 'mrna_len'] = f.mrna_end - f.mrna_start\n",
    "        f.loc[:, 'start'] = f_slice_start\n",
    "        f.loc[:, 'end'] = (f_slice_start + length)\n",
    "        f.loc[:, 'end'] = f[['end', 'mrna_len']].min(axis=1)\n",
    "        f = f[['chromosome', 'geneid', 'transcriptid', 'start', 'end']]\n",
    "        frames.append(f)\n",
    "    intron_edges = pd.concat(frames, axis=0)\n",
    "    intron_edges.loc[:, 'type'] = 'intron-edge'\n",
    "    return intron_edges"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([64, 48])"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| hide\n",
    "example_sample_intron_edges = sample_intron_edges(intron_locations, allowed_intron_edges, random_state=random_state)\n",
    "(example_sample_intron_edges.end - example_sample_intron_edges.start).unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    1.703820e+06\n",
       "mean     6.998604e+03\n",
       "std      2.186718e+04\n",
       "min      1.900000e+01\n",
       "25%      6.170000e+02\n",
       "50%      1.796000e+03\n",
       "75%      4.935000e+03\n",
       "max      1.160411e+06\n",
       "dtype: float64"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| hide\n",
    "intron_lengths = intron_locations.intron_end - intron_locations.intron_start\n",
    "intron_lengths.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(19, 1160411)"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| hide\n",
    "intron_lengths.min(), intron_lengths.max()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def sample_introns(\n",
    "        locations: pd.DataFrame, n: int,\n",
    "        random_state: int = 42, length: int = 64) -> pd.DataFrame:\n",
    "    random.seed(random_state)\n",
    "    \"Get training instances where most of the tokens are <intron>.\"\n",
    "    replace = False if n < locations.shape[0] else True\n",
    "    intron_sample = locations.sample(n, replace=replace, random_state=random_state)\n",
    "    # Handle sequences of varying sizes\n",
    "    intron_sample.loc[:, 'intron_length'] = intron_sample.intron_end - intron_sample.intron_start\n",
    "    intron_len_mask = intron_sample.intron_length <= length\n",
    "    small_introns = intron_sample[intron_len_mask]\n",
    "    large_introns = intron_sample[~intron_len_mask]\n",
    "    sample_frames = []\n",
    "    if small_introns.shape[0] > 0:\n",
    "        # For those introns less than length, center, return the whole thing\n",
    "        # Start at intron start\n",
    "        small_intron_slice_center = small_introns.intron_start\n",
    "        # Shift slice center half the distance of the target sequence\n",
    "        small_intron_slice_center = small_intron_slice_center.subtract(int(length / 2)).apply(lambda val: max(0, val))\n",
    "        small_introns.loc[:, 'start'] = small_intron_slice_center\n",
    "        small_introns.loc[:, 'end'] = small_introns.start + length\n",
    "        small_introns.loc[:, 'end'] = small_introns[['end', 'mrna_end']].min(axis=1)\n",
    "        small_introns = small_introns[['chromosome', 'geneid', 'transcriptid', 'start', 'end']]\n",
    "        small_introns.loc[:, 'type'] = 'intron-small'\n",
    "        sample_frames.append(small_introns)\n",
    "    if large_introns.shape[0] > 0:\n",
    "        # For larger introns, identify the range we can slice to avoid edges\n",
    "        large_introns.loc[:, 'slice_max'] = large_introns.intron_end - length\n",
    "        large_introns.loc[:, 'slice_range'] = large_introns.apply(lambda row: range(row.intron_start, row.slice_max + 1, 1), axis=1)\n",
    "        large_introns.loc[:, 'start'] = large_introns.slice_range.apply(lambda r: random.choice(r))\n",
    "        large_introns.loc[:, 'end'] = large_introns.start + length\n",
    "        large_introns = large_introns[['chromosome', 'geneid', 'transcriptid', 'start', 'end']]\n",
    "        large_introns.loc[:, 'type'] = 'intron'\n",
    "        sample_frames.append(large_introns)\n",
    "    # Randomly select a slice point within the identified range\n",
    "    introns = pd.concat(sample_frames, axis=0)\n",
    "    return introns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>geneid</th>\n",
       "      <th>transcriptid</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1558440</th>\n",
       "      <td>NC_000019.10</td>\n",
       "      <td>GeneID:946</td>\n",
       "      <td>XM_011527538.4</td>\n",
       "      <td>57</td>\n",
       "      <td>121</td>\n",
       "      <td>intron-small</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>612326</th>\n",
       "      <td>NC_000006.12</td>\n",
       "      <td>GeneID:124901227</td>\n",
       "      <td>XM_047419609.1</td>\n",
       "      <td>0</td>\n",
       "      <td>64</td>\n",
       "      <td>intron-small</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>209595</th>\n",
       "      <td>NC_000002.12</td>\n",
       "      <td>GeneID:1768</td>\n",
       "      <td>XM_011532649.3</td>\n",
       "      <td>251297</td>\n",
       "      <td>251361</td>\n",
       "      <td>intron</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>926450</th>\n",
       "      <td>NC_000010.11</td>\n",
       "      <td>GeneID:657</td>\n",
       "      <td>NM_001406576.1</td>\n",
       "      <td>86904</td>\n",
       "      <td>86968</td>\n",
       "      <td>intron</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>584454</th>\n",
       "      <td>NC_000005.10</td>\n",
       "      <td>GeneID:3187</td>\n",
       "      <td>NM_001257293.2</td>\n",
       "      <td>2153</td>\n",
       "      <td>2217</td>\n",
       "      <td>intron</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           chromosome            geneid    transcriptid   start     end   \n",
       "1558440  NC_000019.10        GeneID:946  XM_011527538.4      57     121  \\\n",
       "612326   NC_000006.12  GeneID:124901227  XM_047419609.1       0      64   \n",
       "209595   NC_000002.12       GeneID:1768  XM_011532649.3  251297  251361   \n",
       "926450   NC_000010.11        GeneID:657  NM_001406576.1   86904   86968   \n",
       "584454   NC_000005.10       GeneID:3187  NM_001257293.2    2153    2217   \n",
       "\n",
       "                 type  \n",
       "1558440  intron-small  \n",
       "612326   intron-small  \n",
       "209595         intron  \n",
       "926450         intron  \n",
       "584454         intron  "
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| hide\n",
    "example_sample_introns = sample_introns(intron_locations, int(training_instance_target * proportion_intron), random_state=random_state)\n",
    "example_sample_introns.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([64])"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| hide\n",
    "(example_sample_introns.end - example_sample_introns.start).unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>geneid</th>\n",
       "      <th>transcriptid</th>\n",
       "      <th>intron_start</th>\n",
       "      <th>intron_end</th>\n",
       "      <th>mrna_start</th>\n",
       "      <th>mrna_end</th>\n",
       "      <th>chromosome</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GeneID:79501</td>\n",
       "      <td>NM_001005484.2</td>\n",
       "      <td>15</td>\n",
       "      <td>101</td>\n",
       "      <td>0</td>\n",
       "      <td>6167</td>\n",
       "      <td>NC_000001.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GeneID:79501</td>\n",
       "      <td>NM_001005484.2</td>\n",
       "      <td>155</td>\n",
       "      <td>3618</td>\n",
       "      <td>0</td>\n",
       "      <td>6167</td>\n",
       "      <td>NC_000001.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GeneID:112268260</td>\n",
       "      <td>XM_047436352.1</td>\n",
       "      <td>186</td>\n",
       "      <td>547</td>\n",
       "      <td>0</td>\n",
       "      <td>17102</td>\n",
       "      <td>NC_000001.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GeneID:112268260</td>\n",
       "      <td>XM_047436352.1</td>\n",
       "      <td>1339</td>\n",
       "      <td>2365</td>\n",
       "      <td>0</td>\n",
       "      <td>17102</td>\n",
       "      <td>NC_000001.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GeneID:112268260</td>\n",
       "      <td>XM_047436352.1</td>\n",
       "      <td>2467</td>\n",
       "      <td>8912</td>\n",
       "      <td>0</td>\n",
       "      <td>17102</td>\n",
       "      <td>NC_000001.11</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             geneid    transcriptid  intron_start  intron_end  mrna_start   \n",
       "0      GeneID:79501  NM_001005484.2            15         101           0  \\\n",
       "1      GeneID:79501  NM_001005484.2           155        3618           0   \n",
       "2  GeneID:112268260  XM_047436352.1           186         547           0   \n",
       "3  GeneID:112268260  XM_047436352.1          1339        2365           0   \n",
       "4  GeneID:112268260  XM_047436352.1          2467        8912           0   \n",
       "\n",
       "   mrna_end    chromosome  \n",
       "0      6167  NC_000001.11  \n",
       "1      6167  NC_000001.11  \n",
       "2     17102  NC_000001.11  \n",
       "3     17102  NC_000001.11  \n",
       "4     17102  NC_000001.11  "
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| hide\n",
    "intron_locations.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>geneid</th>\n",
       "      <th>transcriptid</th>\n",
       "      <th>intron_start</th>\n",
       "      <th>intron_end</th>\n",
       "      <th>mrna_start</th>\n",
       "      <th>mrna_end</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>mrna_seq_start</th>\n",
       "      <th>mrna_seq_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>166694</th>\n",
       "      <td>GeneID:10000</td>\n",
       "      <td>NM_001206729.2</td>\n",
       "      <td>41</td>\n",
       "      <td>302</td>\n",
       "      <td>7495</td>\n",
       "      <td>351356</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>7495</td>\n",
       "      <td>7536</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166695</th>\n",
       "      <td>GeneID:10000</td>\n",
       "      <td>NM_001206729.2</td>\n",
       "      <td>460</td>\n",
       "      <td>147868</td>\n",
       "      <td>7495</td>\n",
       "      <td>351356</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>7797</td>\n",
       "      <td>7955</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166696</th>\n",
       "      <td>GeneID:10000</td>\n",
       "      <td>NM_001206729.2</td>\n",
       "      <td>147994</td>\n",
       "      <td>178701</td>\n",
       "      <td>7495</td>\n",
       "      <td>351356</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>155363</td>\n",
       "      <td>155489</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166697</th>\n",
       "      <td>GeneID:10000</td>\n",
       "      <td>NM_001206729.2</td>\n",
       "      <td>178813</td>\n",
       "      <td>197547</td>\n",
       "      <td>7495</td>\n",
       "      <td>351356</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>186196</td>\n",
       "      <td>186308</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166698</th>\n",
       "      <td>GeneID:10000</td>\n",
       "      <td>NM_001206729.2</td>\n",
       "      <td>197692</td>\n",
       "      <td>205842</td>\n",
       "      <td>7495</td>\n",
       "      <td>351356</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>205042</td>\n",
       "      <td>205187</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              geneid    transcriptid  intron_start  intron_end  mrna_start   \n",
       "166694  GeneID:10000  NM_001206729.2            41         302        7495  \\\n",
       "166695  GeneID:10000  NM_001206729.2           460      147868        7495   \n",
       "166696  GeneID:10000  NM_001206729.2        147994      178701        7495   \n",
       "166697  GeneID:10000  NM_001206729.2        178813      197547        7495   \n",
       "166698  GeneID:10000  NM_001206729.2        197692      205842        7495   \n",
       "\n",
       "        mrna_end    chromosome  mrna_seq_start  mrna_seq_end  \n",
       "166694    351356  NC_000001.11            7495          7536  \n",
       "166695    351356  NC_000001.11            7797          7955  \n",
       "166696    351356  NC_000001.11          155363        155489  \n",
       "166697    351356  NC_000001.11          186196        186308  \n",
       "166698    351356  NC_000001.11          205042        205187  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>geneid</th>\n",
       "      <th>transcriptid</th>\n",
       "      <th>intron_start</th>\n",
       "      <th>intron_end</th>\n",
       "      <th>mrna_start</th>\n",
       "      <th>mrna_end</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>mrna_seq_start</th>\n",
       "      <th>mrna_seq_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>166706</th>\n",
       "      <td>GeneID:10000</td>\n",
       "      <td>NM_001206729.2</td>\n",
       "      <td>331261</td>\n",
       "      <td>343798</td>\n",
       "      <td>7495</td>\n",
       "      <td>351356</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>343798</td>\n",
       "      <td>351356</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166602</th>\n",
       "      <td>GeneID:10000</td>\n",
       "      <td>NM_001370074.1</td>\n",
       "      <td>338756</td>\n",
       "      <td>345745</td>\n",
       "      <td>0</td>\n",
       "      <td>351356</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>345745</td>\n",
       "      <td>351356</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166693</th>\n",
       "      <td>GeneID:10000</td>\n",
       "      <td>NM_005465.7</td>\n",
       "      <td>337920</td>\n",
       "      <td>344909</td>\n",
       "      <td>836</td>\n",
       "      <td>351356</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>344909</td>\n",
       "      <td>351356</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166587</th>\n",
       "      <td>GeneID:10000</td>\n",
       "      <td>NM_181690.2</td>\n",
       "      <td>297773</td>\n",
       "      <td>330856</td>\n",
       "      <td>7797</td>\n",
       "      <td>362847</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>330856</td>\n",
       "      <td>362847</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              geneid    transcriptid  intron_start  intron_end  mrna_start   \n",
       "166706  GeneID:10000  NM_001206729.2        331261      343798        7495  \\\n",
       "166602  GeneID:10000  NM_001370074.1        338756      345745           0   \n",
       "166693  GeneID:10000     NM_005465.7        337920      344909         836   \n",
       "166587  GeneID:10000     NM_181690.2        297773      330856        7797   \n",
       "\n",
       "        mrna_end    chromosome  mrna_seq_start  mrna_seq_end  \n",
       "166706    351356  NC_000001.11          343798        351356  \n",
       "166602    351356  NC_000001.11          345745        351356  \n",
       "166693    351356  NC_000001.11          344909        351356  \n",
       "166587    362847  NC_000001.11          330856        362847  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>geneid</th>\n",
       "      <th>transcriptid</th>\n",
       "      <th>intron_start</th>\n",
       "      <th>intron_end</th>\n",
       "      <th>mrna_start</th>\n",
       "      <th>mrna_end</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>mrna_seq_start</th>\n",
       "      <th>mrna_seq_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GeneID:10000</td>\n",
       "      <td>NM_001206729.2</td>\n",
       "      <td>41</td>\n",
       "      <td>302</td>\n",
       "      <td>7495</td>\n",
       "      <td>351356</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>7495</td>\n",
       "      <td>7536</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GeneID:10000</td>\n",
       "      <td>NM_001206729.2</td>\n",
       "      <td>460</td>\n",
       "      <td>147868</td>\n",
       "      <td>7495</td>\n",
       "      <td>351356</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>7797</td>\n",
       "      <td>7955</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GeneID:10000</td>\n",
       "      <td>NM_001206729.2</td>\n",
       "      <td>147994</td>\n",
       "      <td>178701</td>\n",
       "      <td>7495</td>\n",
       "      <td>351356</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>155363</td>\n",
       "      <td>155489</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GeneID:10000</td>\n",
       "      <td>NM_001206729.2</td>\n",
       "      <td>178813</td>\n",
       "      <td>197547</td>\n",
       "      <td>7495</td>\n",
       "      <td>351356</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>186196</td>\n",
       "      <td>186308</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GeneID:10000</td>\n",
       "      <td>NM_001206729.2</td>\n",
       "      <td>197692</td>\n",
       "      <td>205842</td>\n",
       "      <td>7495</td>\n",
       "      <td>351356</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>205042</td>\n",
       "      <td>205187</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>GeneID:10000</td>\n",
       "      <td>NM_001206729.2</td>\n",
       "      <td>205974</td>\n",
       "      <td>228423</td>\n",
       "      <td>7495</td>\n",
       "      <td>351356</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>213337</td>\n",
       "      <td>213469</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>GeneID:10000</td>\n",
       "      <td>NM_001206729.2</td>\n",
       "      <td>228489</td>\n",
       "      <td>229845</td>\n",
       "      <td>7495</td>\n",
       "      <td>351356</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>235918</td>\n",
       "      <td>235984</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>GeneID:10000</td>\n",
       "      <td>NM_001206729.2</td>\n",
       "      <td>229914</td>\n",
       "      <td>270536</td>\n",
       "      <td>7495</td>\n",
       "      <td>351356</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>237340</td>\n",
       "      <td>237409</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>GeneID:10000</td>\n",
       "      <td>NM_001206729.2</td>\n",
       "      <td>270659</td>\n",
       "      <td>279736</td>\n",
       "      <td>7495</td>\n",
       "      <td>351356</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>278031</td>\n",
       "      <td>278154</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>GeneID:10000</td>\n",
       "      <td>NM_001206729.2</td>\n",
       "      <td>279865</td>\n",
       "      <td>290641</td>\n",
       "      <td>7495</td>\n",
       "      <td>351356</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>287231</td>\n",
       "      <td>287360</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>GeneID:10000</td>\n",
       "      <td>NM_001206729.2</td>\n",
       "      <td>290856</td>\n",
       "      <td>297987</td>\n",
       "      <td>7495</td>\n",
       "      <td>351356</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>298136</td>\n",
       "      <td>298351</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>GeneID:10000</td>\n",
       "      <td>NM_001206729.2</td>\n",
       "      <td>298075</td>\n",
       "      <td>331158</td>\n",
       "      <td>7495</td>\n",
       "      <td>351356</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>305482</td>\n",
       "      <td>305570</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>GeneID:10000</td>\n",
       "      <td>NM_001206729.2</td>\n",
       "      <td>331261</td>\n",
       "      <td>343798</td>\n",
       "      <td>7495</td>\n",
       "      <td>351356</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>338653</td>\n",
       "      <td>338756</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>GeneID:10000</td>\n",
       "      <td>NM_001206729.2</td>\n",
       "      <td>331261</td>\n",
       "      <td>343798</td>\n",
       "      <td>7495</td>\n",
       "      <td>351356</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>343798</td>\n",
       "      <td>351356</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>GeneID:10000</td>\n",
       "      <td>NM_001370074.1</td>\n",
       "      <td>31</td>\n",
       "      <td>7797</td>\n",
       "      <td>0</td>\n",
       "      <td>351356</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>0</td>\n",
       "      <td>31</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          geneid    transcriptid  intron_start  intron_end  mrna_start   \n",
       "0   GeneID:10000  NM_001206729.2            41         302        7495  \\\n",
       "1   GeneID:10000  NM_001206729.2           460      147868        7495   \n",
       "2   GeneID:10000  NM_001206729.2        147994      178701        7495   \n",
       "3   GeneID:10000  NM_001206729.2        178813      197547        7495   \n",
       "4   GeneID:10000  NM_001206729.2        197692      205842        7495   \n",
       "5   GeneID:10000  NM_001206729.2        205974      228423        7495   \n",
       "6   GeneID:10000  NM_001206729.2        228489      229845        7495   \n",
       "7   GeneID:10000  NM_001206729.2        229914      270536        7495   \n",
       "8   GeneID:10000  NM_001206729.2        270659      279736        7495   \n",
       "9   GeneID:10000  NM_001206729.2        279865      290641        7495   \n",
       "10  GeneID:10000  NM_001206729.2        290856      297987        7495   \n",
       "11  GeneID:10000  NM_001206729.2        298075      331158        7495   \n",
       "12  GeneID:10000  NM_001206729.2        331261      343798        7495   \n",
       "13  GeneID:10000  NM_001206729.2        331261      343798        7495   \n",
       "14  GeneID:10000  NM_001370074.1            31        7797           0   \n",
       "\n",
       "    mrna_end    chromosome  mrna_seq_start  mrna_seq_end  \n",
       "0     351356  NC_000001.11            7495          7536  \n",
       "1     351356  NC_000001.11            7797          7955  \n",
       "2     351356  NC_000001.11          155363        155489  \n",
       "3     351356  NC_000001.11          186196        186308  \n",
       "4     351356  NC_000001.11          205042        205187  \n",
       "5     351356  NC_000001.11          213337        213469  \n",
       "6     351356  NC_000001.11          235918        235984  \n",
       "7     351356  NC_000001.11          237340        237409  \n",
       "8     351356  NC_000001.11          278031        278154  \n",
       "9     351356  NC_000001.11          287231        287360  \n",
       "10    351356  NC_000001.11          298136        298351  \n",
       "11    351356  NC_000001.11          305482        305570  \n",
       "12    351356  NC_000001.11          338653        338756  \n",
       "13    351356  NC_000001.11          343798        351356  \n",
       "14    351356  NC_000001.11               0            31  "
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| hide\n",
    "intron_locations_sample = intron_locations.sort_values(['chromosome', 'geneid', 'transcriptid']).head(50)\n",
    "intron_locations_sample_grouped = intron_locations_sample.groupby(\n",
    "        ['chromosome', 'geneid', 'transcriptid']\n",
    ")\n",
    "intron_locations_sample.loc[:, 'mrna_seq_start'] = (\n",
    "    intron_locations_sample_grouped.intron_end.shift(1).fillna(0) + \\\n",
    "        intron_locations_sample.mrna_start\n",
    ").astype(int)\n",
    "intron_locations_sample.loc[:, 'mrna_seq_end'] = intron_locations_sample.intron_start + \\\n",
    "        intron_locations_sample.mrna_start\n",
    "display(intron_locations_sample.head())\n",
    "# Add last sequence of mRNA from last intron\n",
    "intron_locations_sample_last_intron = intron_locations_sample_grouped.tail(1)\n",
    "intron_locations_sample_last_intron.loc[:, 'mrna_seq_start'] = intron_locations_sample_last_intron.intron_end\n",
    "intron_locations_sample_last_intron.loc[:, 'mrna_seq_end'] = intron_locations_sample_last_intron.mrna_end\n",
    "display(intron_locations_sample_last_intron.head())\n",
    "intron_locations_sample_mrna_locations = pd.concat(\n",
    "    [\n",
    "        intron_locations_sample,\n",
    "        intron_locations_sample_last_intron\n",
    "    ], axis=0, ignore_index=True\n",
    "    ).sort_values(\n",
    "        ['chromosome', 'geneid', 'transcriptid']).reset_index(drop=True)\n",
    "intron_locations_sample_mrna_locations.head(15)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def get_mrna_locations(locations: pd.DataFrame) -> pd.DataFrame:\n",
    "    \"Get the mrna sequences between introns\"\n",
    "    # Get locations of transcribed dna\n",
    "    locations = locations.copy()\n",
    "    locations_grouped = locations.groupby(\n",
    "            ['chromosome', 'geneid', 'transcriptid']\n",
    "    )\n",
    "    mrna_seq_start = (\n",
    "        locations_grouped.intron_end.shift(1).fillna(0)\n",
    "    ).astype(int)\n",
    "    mrna_seq_start.name = \"start\"\n",
    "    mrna_seq_end = locations.intron_start\n",
    "    mrna_seq_end.name = \"end\"\n",
    "    selected_cols = ['chromosome', 'geneid', 'transcriptid', 'mrna_start', 'mrna_end']\n",
    "    mrna_sequences = pd.concat(\n",
    "        [\n",
    "            locations[selected_cols],\n",
    "            mrna_seq_start,\n",
    "            mrna_seq_end\n",
    "        ], axis=1)\n",
    "    # Add last sequence of mRNA from last intron\n",
    "    last_intron = locations_grouped.tail(1)\n",
    "    last_intron.loc[:, 'start'] = last_intron.intron_end\n",
    "    last_intron.loc[:, 'end'] = last_intron.mrna_end\n",
    "    all_mrna_sequences = pd.concat(\n",
    "        [\n",
    "            mrna_sequences,\n",
    "            last_intron[selected_cols + ['start', 'end']]\n",
    "        ], axis=0, ignore_index=True\n",
    "    ).sort_values(['chromosome', 'geneid', 'transcriptid', 'start']).reset_index(drop=True)\n",
    "    return all_mrna_sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>geneid</th>\n",
       "      <th>transcriptid</th>\n",
       "      <th>mrna_start</th>\n",
       "      <th>mrna_end</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>GeneID:10000</td>\n",
       "      <td>NM_001206729.2</td>\n",
       "      <td>7495</td>\n",
       "      <td>351356</td>\n",
       "      <td>0</td>\n",
       "      <td>41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>GeneID:10000</td>\n",
       "      <td>NM_001206729.2</td>\n",
       "      <td>7495</td>\n",
       "      <td>351356</td>\n",
       "      <td>302</td>\n",
       "      <td>460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>GeneID:10000</td>\n",
       "      <td>NM_001206729.2</td>\n",
       "      <td>7495</td>\n",
       "      <td>351356</td>\n",
       "      <td>147868</td>\n",
       "      <td>147994</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>GeneID:10000</td>\n",
       "      <td>NM_001206729.2</td>\n",
       "      <td>7495</td>\n",
       "      <td>351356</td>\n",
       "      <td>178701</td>\n",
       "      <td>178813</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>GeneID:10000</td>\n",
       "      <td>NM_001206729.2</td>\n",
       "      <td>7495</td>\n",
       "      <td>351356</td>\n",
       "      <td>197547</td>\n",
       "      <td>197692</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     chromosome        geneid    transcriptid  mrna_start  mrna_end   start   \n",
       "0  NC_000001.11  GeneID:10000  NM_001206729.2        7495    351356       0  \\\n",
       "1  NC_000001.11  GeneID:10000  NM_001206729.2        7495    351356     302   \n",
       "2  NC_000001.11  GeneID:10000  NM_001206729.2        7495    351356  147868   \n",
       "3  NC_000001.11  GeneID:10000  NM_001206729.2        7495    351356  178701   \n",
       "4  NC_000001.11  GeneID:10000  NM_001206729.2        7495    351356  197547   \n",
       "\n",
       "      end  \n",
       "0      41  \n",
       "1     460  \n",
       "2  147994  \n",
       "3  178813  \n",
       "4  197692  "
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| hide\n",
    "mrna_locations = get_mrna_locations(intron_locations)\n",
    "mrna_locations.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1832931, 7)"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| hide\n",
    "mrna_locations.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>geneid</th>\n",
       "      <th>transcriptid</th>\n",
       "      <th>intron_start</th>\n",
       "      <th>intron_end</th>\n",
       "      <th>mrna_start</th>\n",
       "      <th>mrna_end</th>\n",
       "      <th>chromosome</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>166694</th>\n",
       "      <td>GeneID:10000</td>\n",
       "      <td>NM_001206729.2</td>\n",
       "      <td>41</td>\n",
       "      <td>302</td>\n",
       "      <td>7495</td>\n",
       "      <td>351356</td>\n",
       "      <td>NC_000001.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166695</th>\n",
       "      <td>GeneID:10000</td>\n",
       "      <td>NM_001206729.2</td>\n",
       "      <td>460</td>\n",
       "      <td>147868</td>\n",
       "      <td>7495</td>\n",
       "      <td>351356</td>\n",
       "      <td>NC_000001.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166696</th>\n",
       "      <td>GeneID:10000</td>\n",
       "      <td>NM_001206729.2</td>\n",
       "      <td>147994</td>\n",
       "      <td>178701</td>\n",
       "      <td>7495</td>\n",
       "      <td>351356</td>\n",
       "      <td>NC_000001.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166697</th>\n",
       "      <td>GeneID:10000</td>\n",
       "      <td>NM_001206729.2</td>\n",
       "      <td>178813</td>\n",
       "      <td>197547</td>\n",
       "      <td>7495</td>\n",
       "      <td>351356</td>\n",
       "      <td>NC_000001.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166698</th>\n",
       "      <td>GeneID:10000</td>\n",
       "      <td>NM_001206729.2</td>\n",
       "      <td>197692</td>\n",
       "      <td>205842</td>\n",
       "      <td>7495</td>\n",
       "      <td>351356</td>\n",
       "      <td>NC_000001.11</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              geneid    transcriptid  intron_start  intron_end  mrna_start   \n",
       "166694  GeneID:10000  NM_001206729.2            41         302        7495  \\\n",
       "166695  GeneID:10000  NM_001206729.2           460      147868        7495   \n",
       "166696  GeneID:10000  NM_001206729.2        147994      178701        7495   \n",
       "166697  GeneID:10000  NM_001206729.2        178813      197547        7495   \n",
       "166698  GeneID:10000  NM_001206729.2        197692      205842        7495   \n",
       "\n",
       "        mrna_end    chromosome  \n",
       "166694    351356  NC_000001.11  \n",
       "166695    351356  NC_000001.11  \n",
       "166696    351356  NC_000001.11  \n",
       "166697    351356  NC_000001.11  \n",
       "166698    351356  NC_000001.11  "
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| hide\n",
    "first_sample_mrna = mrna_locations.iloc[0, :]\n",
    "intron_locations[\n",
    "    (intron_locations.chromosome == first_sample_mrna.chromosome) &\n",
    "    (intron_locations.geneid == first_sample_mrna.geneid) &\n",
    "    (intron_locations.transcriptid == first_sample_mrna.transcriptid)\n",
    "].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def sample_mrna(\n",
    "        mrna_locations: pd.DataFrame, n: int, \n",
    "        random_state: int = 42, length: int = 64) -> pd.DataFrame:\n",
    "    \"Get a sample or mrna sequence locations\"\n",
    "    replace = False if n < mrna_locations.shape[0] else True\n",
    "    mrna_locations = mrna_locations.sample(n, replace=replace, random_state=random_state)\n",
    "    # For small mrna sections, do the same thing we did with the introns\n",
    "    # Handle sequences of varying sizes\n",
    "    mrna_locations.loc[:, 'length'] = mrna_locations.end - mrna_locations.start\n",
    "    mrna_len_mask = mrna_locations.length <= length\n",
    "    small_sequences = mrna_locations[mrna_len_mask]\n",
    "    large_sequences = mrna_locations[~mrna_len_mask]\n",
    "    sample_frames = []\n",
    "    if small_sequences.shape[0] > 0:\n",
    "        # For those introns less than length, center, return the whole thing\n",
    "        # Start at intron start\n",
    "        small_sequences_slice_center = small_sequences.start\n",
    "        # Shift slice center half the distance of the target sequence\n",
    "        small_sequences_slice_center = small_sequences_slice_center.subtract(int(length / 2)).apply(lambda val: max(0, val))\n",
    "        small_sequences.loc[:, 'start'] = small_sequences_slice_center\n",
    "        small_sequences.loc[:, 'end'] = small_sequences.start + length\n",
    "        small_sequences.loc[:, 'end'] = small_sequences[['end', 'mrna_end']].min(axis=1)\n",
    "        small_sequences = small_sequences[['chromosome', 'geneid', 'transcriptid', 'start', 'end']]\n",
    "        small_sequences.loc[:, 'type'] = 'mrna-small'\n",
    "        sample_frames.append(small_sequences)\n",
    "    if large_sequences.shape[0] > 0:\n",
    "        # For larger introns, identify the range we can slice to avoid edges\n",
    "        large_sequences.loc[:, 'slice_max'] = large_sequences.end - length\n",
    "        large_sequences.loc[:, 'slice_range'] = large_sequences.apply(lambda row: range(row.start, row.slice_max + 1, 1), axis=1)\n",
    "        large_sequences.loc[:, 'start'] = large_sequences.slice_range.apply(lambda r: random.choice(r))\n",
    "        large_sequences.loc[:, 'end'] = large_sequences.start + length\n",
    "        large_sequences = large_sequences[['chromosome', 'geneid', 'transcriptid', 'start', 'end']]\n",
    "        large_sequences.loc[:, 'type'] = 'mrna'\n",
    "        sample_frames.append(large_sequences)\n",
    "    # Randomly select a slice point within the identified range\n",
    "    return pd.concat(sample_frames, axis=0, ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>geneid</th>\n",
       "      <th>transcriptid</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_000012.12</td>\n",
       "      <td>GeneID:4033</td>\n",
       "      <td>NM_001366547.2</td>\n",
       "      <td>37025</td>\n",
       "      <td>37089</td>\n",
       "      <td>mrna-small</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_000019.10</td>\n",
       "      <td>GeneID:57153</td>\n",
       "      <td>XM_047439113.1</td>\n",
       "      <td>28988</td>\n",
       "      <td>29052</td>\n",
       "      <td>mrna-small</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_000011.10</td>\n",
       "      <td>GeneID:6833</td>\n",
       "      <td>NM_000352.6</td>\n",
       "      <td>59851</td>\n",
       "      <td>59915</td>\n",
       "      <td>mrna-small</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_000018.10</td>\n",
       "      <td>GeneID:374860</td>\n",
       "      <td>XM_011525664.3</td>\n",
       "      <td>21139</td>\n",
       "      <td>21203</td>\n",
       "      <td>mrna-small</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_000009.12</td>\n",
       "      <td>GeneID:56904</td>\n",
       "      <td>XM_005252098.3</td>\n",
       "      <td>16001</td>\n",
       "      <td>16065</td>\n",
       "      <td>mrna-small</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     chromosome         geneid    transcriptid  start    end        type\n",
       "0  NC_000012.12    GeneID:4033  NM_001366547.2  37025  37089  mrna-small\n",
       "1  NC_000019.10   GeneID:57153  XM_047439113.1  28988  29052  mrna-small\n",
       "2  NC_000011.10    GeneID:6833     NM_000352.6  59851  59915  mrna-small\n",
       "3  NC_000018.10  GeneID:374860  XM_011525664.3  21139  21203  mrna-small\n",
       "4  NC_000009.12   GeneID:56904  XM_005252098.3  16001  16065  mrna-small"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| hide\n",
    "example_sample_mrna = sample_mrna(mrna_locations, int(proportion_mrna * training_instance_target))\n",
    "example_sample_mrna.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def sample_mrna_edges(locations: pd.DataFrame, n: int, random_state: int = 42, length: int = 64) -> pd.DataFrame:\n",
    "    \"Get the beginning and end of mrna\"\n",
    "    locations = locations.drop_duplicates(\n",
    "        ['chromosome', 'geneid', 'transcriptid', 'mrna_start', 'mrna_end']\n",
    "    ).drop(['intron_start', 'intron_end'], axis=1).reset_index(drop=True)\n",
    "    n_start = int(n / 2)\n",
    "    n_end = n - n_start\n",
    "    replace = False if (n_start < locations.shape[0]) or (n_end < locations.shape[0]) else True\n",
    "    mrna_starts = locations.sample(\n",
    "        n_start, replace=replace, random_state=random_state\n",
    "        ).rename({'mrna_start': 'start'}, axis=1)\n",
    "    mrna_starts.loc[:, 'end'] = mrna_starts.start + length\n",
    "    mrna_starts.loc[:, 'end'] = mrna_starts[['mrna_end', 'end']].min(axis=1)\n",
    "    mrna_starts.drop(['mrna_end'], axis=1, inplace=True)\n",
    "    mrna_ends = locations.sample(\n",
    "        n_end, replace=replace, random_state=random_state).rename({'mrna_end': 'end'}, axis=1)\n",
    "    mrna_ends.loc[:, 'start'] = mrna_ends.end - length\n",
    "    mrna_ends.loc[:, 'start'] = mrna_ends[['mrna_start', 'start']].max(axis=1)\n",
    "    mrna_ends.drop(['mrna_start'], axis=1, inplace=True)\n",
    "    sample_edges = pd.concat([mrna_starts, mrna_ends], axis=0, ignore_index=True)\n",
    "    sample_edges = sample_edges[['chromosome', 'geneid', 'transcriptid', 'start', 'end']]\n",
    "    sample_edges.loc[:, 'type'] = 'mrna-edge'\n",
    "    return sample_edges"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>geneid</th>\n",
       "      <th>transcriptid</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_000022.11</td>\n",
       "      <td>GeneID:6525</td>\n",
       "      <td>XM_011530338.3</td>\n",
       "      <td>17200</td>\n",
       "      <td>17264</td>\n",
       "      <td>mrna-edge</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_000006.12</td>\n",
       "      <td>GeneID:8676</td>\n",
       "      <td>XM_047419441.1</td>\n",
       "      <td>10554</td>\n",
       "      <td>10618</td>\n",
       "      <td>mrna-edge</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_000008.11</td>\n",
       "      <td>GeneID:138046</td>\n",
       "      <td>NM_001354315.2</td>\n",
       "      <td>487</td>\n",
       "      <td>551</td>\n",
       "      <td>mrna-edge</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_000005.10</td>\n",
       "      <td>GeneID:80006</td>\n",
       "      <td>NM_001365343.1</td>\n",
       "      <td>0</td>\n",
       "      <td>64</td>\n",
       "      <td>mrna-edge</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_000004.12</td>\n",
       "      <td>GeneID:419</td>\n",
       "      <td>XM_047415699.1</td>\n",
       "      <td>0</td>\n",
       "      <td>64</td>\n",
       "      <td>mrna-edge</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     chromosome         geneid    transcriptid  start    end       type\n",
       "0  NC_000022.11    GeneID:6525  XM_011530338.3  17200  17264  mrna-edge\n",
       "1  NC_000006.12    GeneID:8676  XM_047419441.1  10554  10618  mrna-edge\n",
       "2  NC_000008.11  GeneID:138046  NM_001354315.2    487    551  mrna-edge\n",
       "3  NC_000005.10   GeneID:80006  NM_001365343.1      0     64  mrna-edge\n",
       "4  NC_000004.12     GeneID:419  XM_047415699.1      0     64  mrna-edge"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| hide\n",
    "example_sample_mrna_edges = sample_mrna_edges(intron_locations, int(proportion_mrna_edge * training_instance_target))\n",
    "example_sample_mrna_edges.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def sample_sequences_idx(\n",
    "        n: int, \n",
    "        intron_locations: pd.DataFrame,\n",
    "        mrna_locations: pd.DataFrame,\n",
    "        intron_prop: float, intron_edge_prop: float, \n",
    "        mrna_prop: float, mrna_edge_prop: float,\n",
    "        random_state: int = 42,\n",
    "        length: int = 64) -> pd.DataFrame:\n",
    "    \"Build training dataset from intron locations.\"\n",
    "    intron_sample = sample_introns(\n",
    "        intron_locations, int(n * intron_prop), \n",
    "        random_state=random_state, length=length)\n",
    "    intron_edge_sample = sample_intron_edges(\n",
    "        intron_locations, int(n * intron_edge_prop), \n",
    "        random_state=random_state, length=length)\n",
    "    mrna_sample = sample_mrna(mrna_locations, int(n * mrna_prop), \n",
    "        random_state=random_state, length=length)\n",
    "    mrna_edge_sample = sample_mrna_edges(intron_locations, int(n * mrna_edge_prop),\n",
    "        random_state=random_state, length=length)\n",
    "    sample = pd.concat([\n",
    "        intron_sample,\n",
    "        intron_edge_sample,\n",
    "        mrna_sample,\n",
    "        mrna_edge_sample\n",
    "    ], axis=0, ignore_index=True)\n",
    "    return sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "type\n",
       "intron         125\n",
       "intron-edge    125\n",
       "mrna-edge      125\n",
       "mrna           116\n",
       "mrna-small       9\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| hide\n",
    "example_sample_sequence_idx = sample_sequences_idx(\n",
    "    500,\n",
    "    intron_locations,\n",
    "    mrna_locations,\n",
    "    intron_prop=proportion_intron,\n",
    "    intron_edge_prop=proportion_intron_edge,\n",
    "    mrna_prop=proportion_mrna,\n",
    "    mrna_edge_prop=proportion_mrna_edge,\n",
    "    random_state=random_state,\n",
    ")\n",
    "example_sample_sequence_idx['type'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>geneid</th>\n",
       "      <th>transcriptid</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_000002.12</td>\n",
       "      <td>GeneID:1768</td>\n",
       "      <td>XM_011532649.3</td>\n",
       "      <td>251297</td>\n",
       "      <td>251361</td>\n",
       "      <td>intron</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_000010.11</td>\n",
       "      <td>GeneID:657</td>\n",
       "      <td>NM_001406576.1</td>\n",
       "      <td>86904</td>\n",
       "      <td>86968</td>\n",
       "      <td>intron</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_000005.10</td>\n",
       "      <td>GeneID:3187</td>\n",
       "      <td>NM_001257293.2</td>\n",
       "      <td>2153</td>\n",
       "      <td>2217</td>\n",
       "      <td>intron</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_000004.12</td>\n",
       "      <td>GeneID:84286</td>\n",
       "      <td>XM_017008701.2</td>\n",
       "      <td>4587</td>\n",
       "      <td>4651</td>\n",
       "      <td>intron</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_000005.10</td>\n",
       "      <td>GeneID:2533</td>\n",
       "      <td>NM_018594.2</td>\n",
       "      <td>86359</td>\n",
       "      <td>86423</td>\n",
       "      <td>intron</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     chromosome        geneid    transcriptid   start     end    type\n",
       "0  NC_000002.12   GeneID:1768  XM_011532649.3  251297  251361  intron\n",
       "1  NC_000010.11    GeneID:657  NM_001406576.1   86904   86968  intron\n",
       "2  NC_000005.10   GeneID:3187  NM_001257293.2    2153    2217  intron\n",
       "3  NC_000004.12  GeneID:84286  XM_017008701.2    4587    4651  intron\n",
       "4  NC_000005.10   GeneID:2533     NM_018594.2   86359   86423  intron"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| hide\n",
    "example_sample_sequence_idx.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def get_training_sequences_with_idx(\n",
    "        chromosome: str, geneid: str, transcriptid: str, \n",
    "        genes: pd.DataFrame,\n",
    "        intron_locations: pd.DataFrame,\n",
    "        start: int, end: int,\n",
    "        debug: bool = False\n",
    "        ) -> tuple[str, str, int]:\n",
    "    \"\"\n",
    "    time_start = time.time()\n",
    "    gene_sequence = get_gene(genes, geneid)\n",
    "    read_gene_time = time.time()\n",
    "    intron_locations = intron_locations[\n",
    "        (intron_locations.chromosome == chromosome) &\n",
    "        (intron_locations.geneid == geneid) &\n",
    "        (intron_locations.transcriptid == transcriptid)\n",
    "    ]\n",
    "    mrna_start = intron_locations.iloc[0, :].mrna_start\n",
    "    mrna_end = intron_locations.iloc[0, :].mrna_end\n",
    "    intron_list = get_mrna_intron_locations(\n",
    "        chromosome, geneid, transcriptid, intron_locations\n",
    "    )\n",
    "    mrna_intron_list_time = time.time()\n",
    "    annotated_gene, annotated_mrna = get_mrna_from_gene(\n",
    "        gene_sequence, \n",
    "        mrna_start, mrna_end, \n",
    "        intron_list)\n",
    "    annotation_sequences_time = time.time()\n",
    "    if debug:\n",
    "        print(\"Time to read gene:\", timedelta(seconds=read_gene_time - time_start))\n",
    "        print(\"Time to get list of introns:\", timedelta(seconds=mrna_intron_list_time - read_gene_time))\n",
    "        print(\"Time to annotate sequences:\", timedelta(seconds=annotation_sequences_time - mrna_intron_list_time))\n",
    "        print(\"Total time:\", timedelta(seconds=annotation_sequences_time - time_start))\n",
    "    return annotated_gene[start: end], annotated_mrna[start: end], start"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chromosome        NC_000002.12\n",
      "geneid             GeneID:1768\n",
      "transcriptid    XM_011532649.3\n",
      "start                   251297\n",
      "end                     251361\n",
      "type                    intron\n",
      "Name: 0, dtype: object\n",
      "Time to read gene: 0:00:00.007806\n",
      "Time to get list of introns: 0:00:00.193654\n",
      "Time to annotate sequences: 0:00:00.188481\n",
      "Total time: 0:00:00.389941\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(64, 64, 251297)"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| hide\n",
    "example_idx_row = example_sample_sequence_idx.iloc[0, :]\n",
    "print(example_idx_row)\n",
    "example_training_sequence_gene, example_training_sequence_mrna, example_training_sequence_position = get_training_sequences_with_idx(\n",
    "    example_idx_row.chromosome, example_idx_row.geneid, example_idx_row.transcriptid, \n",
    "    genes,\n",
    "    intron_locations,\n",
    "    example_idx_row.start, example_idx_row.end,\n",
    "    debug=True\n",
    ")\n",
    "len(example_training_sequence_gene), len(example_training_sequence_mrna), example_training_sequence_position"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|| 10/10 [00:02<00:00,  4.17it/s]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0    ([A, A, G, A, G, G, A, A, G, A, A, G, A, A, G,...\n",
       "1    ([A, C, A, A, A, G, A, C, A, T, A, C, C, C, A,...\n",
       "2    ([A, G, A, G, G, G, C, G, T, G, C, T, C, G, G,...\n",
       "3    ([T, A, A, C, C, A, T, C, T, T, T, T, T, T, T,...\n",
       "4    ([A, C, T, G, G, A, T, C, A, C, A, A, A, G, A,...\n",
       "dtype: object"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| hide\n",
    "example_training_data = example_sample_sequence_idx.head(10).progress_apply(\n",
    "    lambda row: get_training_sequences_with_idx(\n",
    "        row.chromosome, row.geneid, row.transcriptid, \n",
    "        genes,\n",
    "        intron_locations,\n",
    "        row.start, row.end\n",
    "    ), axis=1)\n",
    "example_training_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| hide\n",
    "import nbdev; nbdev.nbdev_export()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
